<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>Introduction to Workflows with Common Workflow Language: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" type="text/css" href="../assets/styles.css">
<script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="manifest" href="../site.webmanifest">
<link rel="mask-icon" href="../safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav "><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="" src="../assets/images/-logo.svg"><abbr class="badge badge-light" title="This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team." style="background-color: #001483; border-radius: 5px">
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#polishing-beta-stage" class="external-link alert-link" style="color: #FFF7F1">
            <i aria-hidden="true" class="icon" data-feather="alert-circle" style="border-radius: 5px"></i>
            Beta
          </a>
          <span class="visually-hidden">This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='../aio.html';">Learner View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav " aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="" src="../assets/images/-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      Introduction to Workflows with Common Workflow Language
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            Introduction to Workflows with Common Workflow Language
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><hr></ul>
</li>
      </ul>
</div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a class="btn btn-primary" href="../aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Introduction to Workflows with Common Workflow Language
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress ">
    <div class="progress-bar " role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../aio.html">Learner View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->
      
            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-introduction.html">1. Introduction</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-shell_to_cwl.html">2. CWL and Shell Tools</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-dependency_graphs.html">3. Developing Multi-Step Workflows</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-reusing_tools.html">4. Resources for Reusing Tools and Scripts</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="addendum-01-debug.html">5. Debugging Workflows</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="more_info.html">6. More Information</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr>
</ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width resources">
<a href="../instructor/aio.html">See all in one page</a>
            

            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">
            
            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-01-introduction"><p>Content from <a href="01-introduction.html">Introduction</a></p>
<hr>
<p>Last updated on 2024-05-14 |
        
        <a href="https://github.com/__USERNAME__/__PROJECT__/edit/main/episodes/01-introduction.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 0 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What are computational workflows, and why are they useful?</li>
<li>What is Common Workflow Language?</li>
<li>How are CWL workflows written?</li>
<li>How do CWL workflows compare to shell workflows?</li>
<li>What are the advantages of using CWL workflows?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Understand why you might use CWL instead of a shell script</li>
<li>Recognise the advantages of using a computational workflow</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="computational-workflows"><h2 class="section-heading">Computational Workflows<a class="anchor" aria-label="anchor" href="#computational-workflows"></a>
</h2>
<hr class="half-width">
<p>A computational workflow is a formalised chain of software tools,
which explicitly defines how data is input, how it flows between tools
and how it is output. Computational workflows are widely used for data
analysis and enable rapid innovation and decision making.</p>
<p>The most important benefit of computational workflows is that they
require a user to write down, fully formalise and automate their
dataflow and process. While this can be a challenge, it allows greater
repeatability, shareability and robustness to be achieved.</p>
<p>Below is an image showing a graphical representation of a dummy
workflow. You can see inputs being made on the left, data flowing and
being processed by tools and passed to the output.</p>
<figure><img src="../fig/dummy_workflow.png" alt="dummy workflow image" class="figure mx-auto d-block"></figure><p>While you can imagine creating shell scripts (Bash or Make) to meet
this need, using a formal workflow language (such as CWL) brings several
further benefits such as introducing abstraction and improved
scalability and portability. We will discuss some of these benefits
here.</p>
<p>Computational workflows explicitly create a divide between a user‚Äôs
dataflow and the computational details which underpin the chain of
tools, placing these elements in separate files. The dataflow is
described by the <strong>workflow</strong>, where tools are described at
a high level. The tools implementation is specified by <strong>tool
descriptors</strong> where the full complexity of a tools usage is
handled. The image below shows how tool descriptors underpin workflow
steps and hide complexity.</p>
<figure><img src="../fig/dummy_workflow_tooldescriptors.png" alt="Dummy workflow tool descriptor" class="figure mx-auto d-block"></figure><p>This abstraction allows the use of heterogeneous tools, potentially
shared by third parties, and allows workflow users to connect and
utilise a wide range of tools and techniques without the need for
significant computational experience. An example of the strength of this
approach is the ability for workflows to use (e.g.¬†Docker) containers
‚Äúunder the hood‚Äù without the user needing to install, download or learn
any further technologies.</p>
<p>By adapting tool descriptors to multiple platforms, this abstraction
allows the same workflow to be used on different platforms,
transparently to the workflow user. This means users can move between
local development and cloud and HPC solutions seamlessly.</p>
</section><section id="benefits-of-computational-workflows"><h2 class="section-heading">Benefits of Computational Workflows<a class="anchor" aria-label="anchor" href="#benefits-of-computational-workflows"></a>
</h2>
<hr class="half-width">
<p>In summary, computational workflows bring many benefits and an ideal
computational workflow adopts and provides the properties below:</p>
<p><strong>Handy Properties of Computational Workflows</strong><a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a></p>
<table class="table">
<colgroup>
<col width="19%">
<col width="13%">
<col width="66%">
</colgroup>
<tbody>
<tr class="odd">
<td><img src="../fig/PropsCompute01.png" alt="props compute 01 png" class="figure"></td>
<td><strong>Composition &amp; Abstraction</strong></td>
<td>Using the best code written by 3rd parties<br>Handle
heterogeneity<br>Shield Complexity &amp; incompatibility<br>Shareable
reusable, re-mixable methods</td>
</tr>
<tr class="even">
<td><img src="../fig/PropsCompute02.png" alt="props compute 02 png" class="figure"></td>
<td><strong>Sharing &amp; Adaptability</strong></td>
<td>Shared method, publishable know-how<br>BYOD /
parameters<br>Different implementations<br>Changes in execution
infrastructure</td>
</tr>
<tr class="odd">
<td><img src="../fig/PropsCompute03.png" alt="props compute 03 png" class="figure"></td>
<td><strong>Automation</strong></td>
<td>Repetitive reproducible pipelines<br>Simulation sweeps<br>Manage
data and control flow<br>Optimised monitoring &amp;
recovery<br>Automated deployment</td>
</tr>
<tr class="even">
<td><img src="../fig/PropsCompute04.png" alt="props compute 04 png" class="figure"></td>
<td><strong>Reporting &amp; Accreditation</strong></td>
<td>Provenance logging &amp; data
lineage<br>Auto-documentation<br>Result comparison</td>
</tr>
<tr class="odd">
<td><img src="../fig/PropsCompute05.png" alt="props compute 05 png" class="figure"></td>
<td><strong>Scalability &amp; Infrastructure Access</strong></td>
<td>Accessing infrastructures, datasets and tools <br> Optimised
computation and data handling <br> Parallelisation <br> Secure sensitive
data access &amp; management <br> Interoperating datasets &amp;
permission handling</td>
</tr>
<tr class="even">
<td><img src="../fig/PropsCompute06.png" alt="props compute 06 png" class="figure"></td>
<td><strong>Portability</strong></td>
<td>Dependency handling <br> Containerisation &amp; packaging <br>
Moving between on premise &amp; cloud</td>
</tr>
</tbody>
</table></section><section id="computational-workflow-managers"><h2 class="section-heading">Computational Workflow Managers<a class="anchor" aria-label="anchor" href="#computational-workflow-managers"></a>
</h2>
<hr class="half-width">
<p>Computational workflow managers further extend this abstraction,
providing high level tools for managing data and tools, aiming to help
users to design and run computational workflows more easily. A
computational workflow engine provides an interface for launching
workflows, specifying and handling inputs and collecting and exporting
outputs, they can also help users by storing completed steps of
workflows, allowing workflows to be resumed part way, or rerun with
minimal changes.</p>
<p>Further, computational workflow managers aid users with the
automation, monitoring and provenance tracking of the dataflow. They may
also help users to produce and understand reports and outputs from their
workflow.</p>
<div id="workflows-and-provenance" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="workflows-and-provenance" class="callout-inner">
<h3 class="callout-title">Workflows and Provenance<a class="anchor" aria-label="anchor" href="#workflows-and-provenance"></a>
</h3>
<div class="callout-content">
<p>Another advantage of workflows is that by producing computational
workflows in a standard format, and publishing them (alongside any data)
with open access, allow dataflows to be shared in a more FAIR (Findable,
Accessible, Interoperable, and Reusable) manner.</p>
</div>
</div>
</div>
</section><section id="why-a-standard-for-workflows-is-needed"><h2 class="section-heading">Why a standard for workflows is needed?<a class="anchor" aria-label="anchor" href="#why-a-standard-for-workflows-is-needed"></a>
</h2>
<hr class="half-width">
<p>The rise in popularity of workflows has been matched by a rise in the
number of disparate workflow managers that are available, each with
their own syntax or methods for describing the tools and workflows,
reducing portability and interoperability of these workflows. For a
comprehensive lists of all known computational workflow systems, see <a href="https://github.com/common-workflow-language/common-workflow-language/wiki/Existing-Workflow-systems" class="external-link">Computational
Data Analysis Workflow Systems</a> maintained by the CWL community. The
Common Workflow Language <a href="https://www.commonwl.org" class="external-link">CWL</a>
standard has been developed to address these problems, and to serve as a
open standard for describing how to run commandline tools and connect
them to create workflows.</p>
</section><section id="common-workflow-language"><h2 class="section-heading">Common Workflow Language<a class="anchor" aria-label="anchor" href="#common-workflow-language"></a>
</h2>
<hr class="half-width">
<p>CWL is a free and open standard for describing command-line tool
based workflows<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>.</p>
<p>CWL provides a common, but reduced, set of abstractions that are both
used in practice and implemented in many popular workflow managers. The
CWL language is declarative, enabling computational workflows to be
constructed from diverse software tools, executing each through their
command-line interface.</p>
<div id="shell-scripts-reduce-portability" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="shell-scripts-reduce-portability" class="callout-inner">
<h3 class="callout-title">Shell scripts reduce portability<a class="anchor" aria-label="anchor" href="#shell-scripts-reduce-portability"></a>
</h3>
<div class="callout-content">
<p>Previously researchers might write shell scripts to link together
these command-line tools. Although these scripts might provide a direct
means of accessing the tools, writing and maintaining them requires
specific knowledge of the system that they will be used on. Shell
scripts are not easily portable, and so researchers can easily end up
spending more time maintaining the scripts than carrying out their
research. The aim of CWL is to reduce that barrier of usage of these
tools to researchers.</p>
</div>
</div>
</div>
<p>CWL workflows are written in a subset of <a href="https://www.commonwl.org/user_guide/topics/yaml-guide.html" class="external-link">YAML</a>,
with a syntax that does not restrict the amount of detail provided for a
tool or workflow. The execution model is explicit, all required elements
of a tool‚Äôs runtime environment must be specified by the CWL
tool-description author. On top of these basic requirements they can
also add hints or requirements to the tool-description, helping to guide
users (and workflow engines) on what resources are needed for a
tool.</p>
<div id="containerisation" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="containerisation" class="callout-inner">
<h3 class="callout-title">Containerisation<a class="anchor" aria-label="anchor" href="#containerisation"></a>
</h3>
<div class="callout-content">
<p>The CWL standard explicitly support the use of software container
technologies, such as <a href="https://www.docker.com/" class="external-link">docker</a>
helping ensure that the execution of tools is reproducible.</p>
</div>
</div>
</div>
<p>Data locations are explicitly defined, and working directories are
kept separate for each tool invocation. This ensures the portability of
tools and workflows, allowing the same workflows to be run on your local
machine, or in a HPC or cloud environment, with minimal changes
required.</p>
</section><section id="rna-sequencing-example"><h2 class="section-heading">RNA sequencing example<a class="anchor" aria-label="anchor" href="#rna-sequencing-example"></a>
</h2>
<hr class="half-width">
<p>In this tutorial a bioinformatics RNA-sequencing analysis is used as
an example. However, there is no specific knowledge needed for this
tutorial. RNA-sequencing is a technique which examines the quantity and
sequences of <a href="https://en.wikipedia.org/wiki/RNA" class="external-link">RNA</a> in a
sample using next-generation sequencing. The RNA reads are analyzed to
quantify the relative abundance of different RNA molecules in the
sample, a process known as differential gene expression analysis.</p>
<p>The process looks like this:</p>
<figure><img src="../fig/RNAseqWorkflow.png" style='height: "200px"' alt="RNASeq Workflow graph" class="figure mx-auto d-block"><div class="figcaption">Diagram showing a typical RNA sequencing
workflow. The workflow is linear, starting from taking biological
samples and sequence reads, through quality control and trimming steps,
to mapping to a genome and counting gene reads, to finally carrying out
statistical analysis to identify differentially expressed genes.</div>
</figure><p>During this tutorial, the following analytical steps will be
performed.</p>
<ul>
<li>Quality control (FASTQC)</li>
<li>Adapter trimming</li>
<li>Alignment (mapping)</li>
<li>Counting reads associated with genes</li>
</ul>
<p>The different tools necessary for this analysis are already
available. In this tutorial a workflow will be set up to connect these
tools and generate the desired output files.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>CWL is a standard for describing workflows based on command-line
tools.</li>
<li>CWL workflows are written in a subset of YAML.</li>
<li>A CWL workflow is more portable than a shell script.</li>
<li>CWL supports software containers, supporting reproducibility on
different machines.</li>
</ul>
</div>
</div>
</div>
</section><section id="references"><h2 class="section-heading">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<hr class="half-width">
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 --></section><div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p>C. Goble (2021): FAIR Computational Workflows. JOBIM
Proceedings. <a href="https://www.slideshare.net/carolegoble/fair-computational-workflows-249721518" class="external-link">https://www.slideshare.net/carolegoble/fair-computational-workflows-249721518</a><a href="#fnref1" class="footnote-back">‚Ü©Ô∏é</a></p></li>
<li id="fn2"><p>M. Wilkinson, M. Dumontier, I. Aalbersberg, et
al.¬†(2016): The FAIR Guiding Principles for scientific data management
and stewardship. Scientific Data. <a href="https://doi.org/10.1038/sdata.2016.18" class="external-link">https://doi.org/10.1038/sdata.2016.18</a><a href="#fnref2" class="footnote-back">‚Ü©Ô∏é</a></p></li>
</ol>
</div></section><section id="aio-02-shell_to_cwl"><p>Content from <a href="02-shell_to_cwl.html">CWL and Shell Tools</a></p>
<hr>
<p>Last updated on 2024-05-14 |
        
        <a href="https://github.com/__USERNAME__/__PROJECT__/edit/main/episodes/02-shell_to_cwl.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 0 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is the difference between a CWL tool description and a CWL
workflow?</li>
<li>How can we create a tool descriptor?</li>
<li>How can we use this in a single step workflow?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>describe the relationship between a tool and its corresponding CWL
document</li>
<li>exercise good practices when naming inputs and outputs</li>
<li>understand how to reference files for input and output</li>
<li>explain that only files explicitly mentioned in a description will
be included in the output of a step/workflow</li>
<li>implement bulk capturing of all files produced by a step/workflow
for debugging purposes</li>
<li>use STDIN and STDOUT as input and output</li>
<li>capture output written to a specific directory, the working
directory, or the same directory where input is located</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<p>CWL workflows are written in the YAML syntax. This <a href="https://www.commonwl.org/user_guide/topics/yaml-guide.html" class="external-link">short
tutorial</a> explains the parts of YAML used in CWL. A CWL document
contains the workflow and the requirements for running that workflow.
All CWL documents should start with two lines of code:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">YAML<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode yaml" tabindex="0"><code class="sourceCode yaml"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">cwlVersion</span><span class="kw">:</span><span class="at"> v1.2</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">class</span><span class="kw">:</span></span></code></pre>
</div>
<p>The <code>cwlVersion</code> string defines which standard of the
language is required for the tool or workflow. The most recent version
is v1.2.</p>
<p>The <code>class</code> field defines what this particular document
is. The majority of CWL documents will fall into one of two classes:
<code>CommandLineTool</code>, or <code>Workflow</code>. The
<code>CommandLineTool</code> class is used for describing the interface
for a command-line tool, while the <code>Workflow</code> class is used
for connecting those tool descriptions into a workflow. In this lesson
the differences between these two classes are explained, how to pass
data to and from command-line tools and specify working environments for
these, and finally how to use a tool description within a workflow.</p>
<p>You should follow the examples in this lesson from your
<code>novice-tutorial-exercises</code> directory.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">$</span> cd novice-tutorial-exercises</span></code></pre>
</div>
<section id="our-first-cwl-commandlinetool"><h2 class="section-heading">Our first CWL CommandLineTool<a class="anchor" aria-label="anchor" href="#our-first-cwl-commandlinetool"></a>
</h2>
<hr class="half-width">
<p>To demonstrate the basic requirements for a tool descriptor a CWL
description for the popular ‚ÄúHello world!‚Äù demonstration will be
examined.</p>
<p><strong>echo.cwl</strong></p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">YAML<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode yaml" tabindex="0"><code class="sourceCode yaml"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">cwlVersion</span><span class="kw">:</span><span class="at"> v1.2</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">class</span><span class="kw">:</span><span class="at"> CommandLineTool</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="fu">baseCommand</span><span class="kw">:</span><span class="at"> echo</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="fu">inputs</span><span class="kw">:</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="at">  </span><span class="fu">message_text</span><span class="kw">:</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> string</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="at">    </span><span class="fu">inputBinding</span><span class="kw">:</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="at">      </span><span class="fu">position</span><span class="kw">:</span><span class="at"> </span><span class="dv">1</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a></span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a><span class="fu">outputs</span><span class="kw">:</span><span class="at"> </span><span class="kw">[]</span></span></code></pre>
</div>
<p>Next, the input file to the command line tool:
<code>hello_world.yml</code>.</p>
<p><strong>hello_world.yml</strong></p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">YAML<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode yaml" tabindex="0"><code class="sourceCode yaml"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">message_text</span><span class="kw">:</span><span class="at"> Hello world!</span></span></code></pre>
</div>
<p>We will use the reference CWL runner, <code>cwltool</code> to run
this CWL document (the <code>.cwl</code> workflow file) along with the
<code>.yml</code> input file.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">$</span> cwltool echo.cwl hello_world.yml</span></code></pre>
</div>
<div id="accordionSpoiler1" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler1" aria-expanded="false" aria-controls="collapseSpoiler1">
  <h3 class="accordion-header" id="headingSpoiler1">
<div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div>Tool Output</h3>
</button>
<div id="collapseSpoiler1" class="accordion-collapse collapse" aria-labelledby="headingSpoiler1" data-bs-parent="#accordionSpoiler1">
<div class="accordion-body">
<pre class="text"><code>INFO Resolved 'echo.cwl' to 'file:///.../echo.cwl'
INFO [job echo.cwl] /private/tmp/docker_tmprm65mucw$ echo \
    'Hello world!'
Hello world!
INFO [job echo.cwl] completed success
{}
INFO Final process status is success</code></pre>
</div>
</div>
</div>
</div>
<p>The output displayed above shows that the program has run
successfully and its output, <code>Hello world!</code>.</p>
<p>Let‚Äôs take a look at the <code>echo.cwl</code> script in more
detail.</p>
<div class="section level3">
<h3 id="the-commandlinetool-skeleton">The CommandLineTool Skeleton<a class="anchor" aria-label="anchor" href="#the-commandlinetool-skeleton"></a>
</h3>
<p>As explained above, the first 2 lines are always the same, the CWL
version and the class of the script are defined.</p>
<p>In this example the class value is <code>CommandLineTool</code>.</p>
<p>For a <em>CommandLineTool</em> the <code>baseCommand</code>, contains
the command that will be run <code>echo</code>.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">YAML<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode yaml" tabindex="0"><code class="sourceCode yaml"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">inputs</span><span class="kw">:</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="at">  </span><span class="fu">message_text</span><span class="kw">:</span></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> string</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="at">    </span><span class="fu">inputBinding</span><span class="kw">:</span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="at">       </span><span class="fu">position</span><span class="kw">:</span><span class="at"> </span><span class="dv">1</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="inputs">Inputs<a class="anchor" aria-label="anchor" href="#inputs"></a>
</h3>
<p>This block of code contains the <code>inputs</code> section of the
tool description. This section provides the inputs that are needed for
running this specific tool.</p>
<p>To run this example we can provide the input on the command line. In
this case, we have only one input called <code>message_text</code> and
it is of type <code>string</code>.</p>
<div id="input-binding" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="input-binding" class="callout-inner">
<h3 class="callout-title">Input Binding<a class="anchor" aria-label="anchor" href="#input-binding"></a>
</h3>
<div class="callout-content">
<p>A CommandLineTool will run a command. The <code>inputBinding</code>
field is used to define how the input is passed to the command as
parameters.</p>
<p>Here the <code>position</code> field indicates at which position the
input will be on the command line; in this case the
<code>message_text</code> value will be the first thing added to the
command line (after the <code>baseCommand</code>,
<code>echo</code>).</p>
<p>Another common inputBinding attribute is <code>prefix</code> for
adding a prefix to the input value like <code>--arg</code>. If the input
value is not prefixed, the <code>prefix</code> attribute is not attached
to the command line string.</p>
<p>Arguments that are required for a command line tool should be added
to the <code>arguments</code> section</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="outputs">Outputs<a class="anchor" aria-label="anchor" href="#outputs"></a>
</h3>
<p>Lastly the <code>outputs</code> of the tool description. This example
doesn‚Äôt have a formal output. The text is printed directly in the
terminal. So an empty YAML list (<code>[]</code>) is used as the
output.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">YAML<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode yaml" tabindex="0"><code class="sourceCode yaml"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="fu">outputs</span><span class="kw">:</span><span class="at"> </span><span class="kw">[]</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="semantics">Semantics<a class="anchor" aria-label="anchor" href="#semantics"></a>
</h3>
<div id="yaml-key-orders" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="yaml-key-orders" class="callout-inner">
<h3 class="callout-title">YAML Key Orders<a class="anchor" aria-label="anchor" href="#yaml-key-orders"></a>
</h3>
<div class="callout-content">
<p>To make the script more readable the <code>input</code> field is put
in front of the <code>output</code> field. However, CWL syntax requires
only that each field is properly defined, it does not require them to be
in a particular order.</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="challenges">Challenges<a class="anchor" aria-label="anchor" href="#challenges"></a>
</h3>
<div id="changing-the-output-string" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="changing-the-output-string" class="callout-inner">
<h3 class="callout-title">Changing the output string üå∂<a class="anchor" aria-label="anchor" href="#changing-the-output-string"></a>
</h3>
<div class="callout-content">
<p>What do you need to change to print a different text on the command
line?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">Show me the solution</h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>To change the text on the command line, you only have to change the
text in the <code>hello_world.yml</code> file.</p>
<p>For example</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">YAML<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode yaml" tabindex="0"><code class="sourceCode yaml"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">message_text</span><span class="kw">:</span><span class="at"> Good job!</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="updating-the-arguments" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="updating-the-arguments" class="callout-inner">
<h3 class="callout-title">Updating the arguments üå∂üå∂<a class="anchor" aria-label="anchor" href="#updating-the-arguments"></a>
</h3>
<div class="callout-content">
<p>How can one add the <code>-e</code> argument to the echo command to
interpret backslashes?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">Show me the solution</h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">YAML<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode yaml" tabindex="0"><code class="sourceCode yaml"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="fu">baseCommand</span><span class="kw">:</span><span class="at"> echo</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a><span class="fu">arguments</span><span class="kw">:</span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="at">  </span><span class="kw">-</span><span class="at"> </span><span class="fu">position</span><span class="kw">:</span><span class="at"> </span><span class="dv">0</span><span class="at">  </span></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a><span class="at">    </span><span class="fu">valueFrom</span><span class="kw">:</span><span class="at"> </span><span class="st">"-e"</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="redirecting-cwltool-stdout-and-stderr" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="redirecting-cwltool-stdout-and-stderr" class="callout-inner">
<h3 class="callout-title">Redirecting cwltool stdout and stderr üå∂üå∂<a class="anchor" aria-label="anchor" href="#redirecting-cwltool-stdout-and-stderr"></a>
</h3>
<div class="callout-content">
<p>Rerun the <code>echo.cwl</code> script but point stdout and stderr to
different files.</p>
<p>What is the difference between the stdout and stderr from the
<code>echo.cwl</code> script?</p>
</div>
</div>
</div>
<div id="accordionHint1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint1" aria-expanded="false" aria-controls="collapseHint1">
  <h4 class="accordion-header" id="headingHint1">Hint: Redirecting CLI stdout and stderr</h4>
</button>
<div id="collapseHint1" class="accordion-collapse collapse" aria-labelledby="headingHint1" data-bs-parent="#accordionHint1">
<div class="accordion-body">
<p>Use the redirectors <code>1&gt;</code> and <code>2&gt;</code> to
redirect stdout and stderr to different files respectively</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">Show me the solution</h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="ex">$</span> cwltool echo.cwl hello_world.yml <span class="dv">1</span><span class="op">&gt;</span>echo_stdout.txt <span class="dv">2</span><span class="op">&gt;</span>echo_stderr.txt</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="specifying-the-stdout-of-the-tool-as-a-cwl-output" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="specifying-the-stdout-of-the-tool-as-a-cwl-output" class="callout-inner">
<h3 class="callout-title">Specifying the stdout of the tool as a CWL
output üå∂üå∂üå∂<a class="anchor" aria-label="anchor" href="#specifying-the-stdout-of-the-tool-as-a-cwl-output"></a>
</h3>
<div class="callout-content">
<p>Using <a href="https://cwl-for-eo.github.io/guide/how-to/cwl-how-to/02-stdout/capture-stdout/" class="external-link">this
tutorial</a> as a guide</p>
<p>Use the <code>stdout</code> field in the <code>outputs</code> section
of the tool description.</p>
<p>The output id should be ‚Äòmessage_out‚Äô and the type of the output
should be ‚Äòstring‚Äô.</p>
<p>You will need to add in the InlineJavascriptRequirement to the
requirements section of the tool description.</p>
<p>How does this change the output of the cwltool command?</p>
</div>
</div>
</div>
<div id="accordionHint2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint2" aria-expanded="false" aria-controls="collapseHint2">
  <h4 class="accordion-header" id="headingHint2">Hint - stdout output binding</h4>
</button>
<div id="collapseHint2" class="accordion-collapse collapse" aria-labelledby="headingHint2" data-bs-parent="#accordionHint2">
<div class="accordion-body">
<p>Copy the ‚ÄòoutputBinding‚Äô from the tutorial ‚Äòverbatim‚Äô.</p>
<p>Your output key should look like this</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">YAML<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode yaml" tabindex="0"><code class="sourceCode yaml"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">outputs</span><span class="kw">:</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="at">  </span><span class="fu">message_out</span><span class="kw">:</span></span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> string</span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a><span class="at">    </span><span class="fu">outputBinding</span><span class="kw">:</span></span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a><span class="at">      </span><span class="fu">glob</span><span class="kw">:</span><span class="at"> message</span></span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a><span class="at">      </span><span class="fu">loadContents</span><span class="kw">:</span><span class="at"> </span><span class="ch">true</span></span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a><span class="at">      </span><span class="fu">outputEval</span><span class="kw">:</span><span class="at"> $( self[0].contents )</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4">Show me the solution</h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">YAML<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode yaml" tabindex="0"><code class="sourceCode yaml"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="fu">cwlVersion</span><span class="kw">:</span><span class="at"> v1.2</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="fu">class</span><span class="kw">:</span><span class="at"> CommandLineTool</span></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a><span class="fu">requirements</span><span class="kw">:</span></span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a><span class="at">  </span><span class="fu">InlineJavascriptRequirement</span><span class="kw">:</span><span class="at"> </span><span class="kw">{}</span></span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a></span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a><span class="fu">baseCommand</span><span class="kw">:</span><span class="at"> echo</span></span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a></span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a><span class="fu">inputs</span><span class="kw">:</span></span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a><span class="at">  </span><span class="fu">message_text</span><span class="kw">:</span></span>
<span id="cb13-11"><a href="#cb13-11" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> string</span></span>
<span id="cb13-12"><a href="#cb13-12" tabindex="-1"></a><span class="at">    </span><span class="fu">inputBinding</span><span class="kw">:</span></span>
<span id="cb13-13"><a href="#cb13-13" tabindex="-1"></a><span class="at">      </span><span class="fu">position</span><span class="kw">:</span><span class="at"> </span><span class="dv">1</span></span>
<span id="cb13-14"><a href="#cb13-14" tabindex="-1"></a></span>
<span id="cb13-15"><a href="#cb13-15" tabindex="-1"></a><span class="fu">stdout</span><span class="kw">:</span><span class="at"> message</span></span>
<span id="cb13-16"><a href="#cb13-16" tabindex="-1"></a></span>
<span id="cb13-17"><a href="#cb13-17" tabindex="-1"></a><span class="fu">outputs</span><span class="kw">:</span></span>
<span id="cb13-18"><a href="#cb13-18" tabindex="-1"></a><span class="at">  </span><span class="fu">message_out</span><span class="kw">:</span></span>
<span id="cb13-19"><a href="#cb13-19" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> string</span></span>
<span id="cb13-20"><a href="#cb13-20" tabindex="-1"></a><span class="at">    </span><span class="fu">outputBinding</span><span class="kw">:</span></span>
<span id="cb13-21"><a href="#cb13-21" tabindex="-1"></a><span class="at">      </span><span class="fu">glob</span><span class="kw">:</span><span class="at"> message</span></span>
<span id="cb13-22"><a href="#cb13-22" tabindex="-1"></a><span class="at">      </span><span class="fu">loadContents</span><span class="kw">:</span><span class="at"> </span><span class="ch">true</span></span>
<span id="cb13-23"><a href="#cb13-23" tabindex="-1"></a><span class="at">      </span><span class="fu">outputEval</span><span class="kw">:</span><span class="at"> $( self[0].contents )</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section id="cwl-single-step-workflow"><h2 class="section-heading">CWL single step workflow<a class="anchor" aria-label="anchor" href="#cwl-single-step-workflow"></a>
</h2>
<hr class="half-width">
<p>The RNA-seq data from the introduction episode will be used for the
first CWL workflow. The first step of RNA-sequencing analysis is a
quality control of the RNA reads using the <code>fastqc</code> tool.
This tool is already available to use so there is no need to write a new
CWL tool description.</p>
<p>This is the workflow file (<code>rna_seq_workflow_1.cwl</code>).</p>
<div class="section level3">
<h3 id="rna_seq_workflow_1-cwl">rna_seq_workflow_1.cwl<a class="anchor" aria-label="anchor" href="#rna_seq_workflow_1-cwl"></a>
</h3>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">YAML<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode yaml" tabindex="0"><code class="sourceCode yaml"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="fu">cwlVersion</span><span class="kw">:</span><span class="at"> v1.2</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="fu">class</span><span class="kw">:</span><span class="at"> Workflow</span></span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a><span class="fu">inputs</span><span class="kw">:</span></span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a><span class="at">  </span><span class="fu">rna_reads_fruitfly</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a></span>
<span id="cb14-7"><a href="#cb14-7" tabindex="-1"></a><span class="fu">steps</span><span class="kw">:</span></span>
<span id="cb14-8"><a href="#cb14-8" tabindex="-1"></a><span class="at">  </span><span class="fu">quality_control</span><span class="kw">:</span></span>
<span id="cb14-9"><a href="#cb14-9" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span><span class="at"> bio-cwl-tools/fastqc/fastqc_2.cwl</span></span>
<span id="cb14-10"><a href="#cb14-10" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb14-11"><a href="#cb14-11" tabindex="-1"></a><span class="at">      </span><span class="fu">reads_file</span><span class="kw">:</span><span class="at"> rna_reads_fruitfly</span></span>
<span id="cb14-12"><a href="#cb14-12" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at">html_file</span><span class="kw">]</span></span>
<span id="cb14-13"><a href="#cb14-13" tabindex="-1"></a></span>
<span id="cb14-14"><a href="#cb14-14" tabindex="-1"></a><span class="fu">outputs</span><span class="kw">:</span></span>
<span id="cb14-15"><a href="#cb14-15" tabindex="-1"></a><span class="at">  </span><span class="fu">quality_report</span><span class="kw">:</span></span>
<span id="cb14-16"><a href="#cb14-16" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb14-17"><a href="#cb14-17" tabindex="-1"></a><span class="at">    </span><span class="fu">outputSource</span><span class="kw">:</span><span class="at"> quality_control/html_file</span></span></code></pre>
</div>
<p>In a <strong>workflow</strong> the fields <code>inputs</code>,
<code>steps</code> and <code>outputs</code> must always be present.</p>
<p>The workflow tasks or steps that you want to run are listed in this
field. At the moment the workflow only contains one step:
<code>quality_control</code>. In the next episodes more steps will be
added to the workflow.</p>
</div>
<div class="section level3">
<h3 id="workflow-inputs">Workflow Inputs<a class="anchor" aria-label="anchor" href="#workflow-inputs"></a>
</h3>
<p>Let‚Äôs take a closer look at the workflow inputs section:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">YAML<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode yaml" tabindex="0"><code class="sourceCode yaml"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="fu">inputs</span><span class="kw">:</span></span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="at">  </span><span class="fu">rna_reads_fruitfly</span><span class="kw">:</span><span class="at"> File</span></span></code></pre>
</div>
<p>We have one variable <code>rna_reads_fruitfly</code> and it has
<code>File</code> as its type.</p>
<p>This input is used in the step <em>quality_control</em> by the fastqc
tool.</p>
<p>In this example the fastq file consists of <em>Drosophila
melanogaster</em> RNA reads.</p>
<div id="input-and-output-names" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="input-and-output-names" class="callout-inner">
<h3 class="callout-title">Input and output names<a class="anchor" aria-label="anchor" href="#input-and-output-names"></a>
</h3>
<div class="callout-content">
<p>To make this workflow interpretable for other researchers,
self-explanatory and sensible variable names are used.</p>
<p>It is very important to give inputs and outputs a sensible name. Try
not to use variable names like <code>inputA</code> or
<code>inputB</code> because others might not understand what is meant by
it</p>
</div>
</div>
</div>
<div id="input-types" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="input-types" class="callout-inner">
<h3 class="callout-title">Input Types<a class="anchor" aria-label="anchor" href="#input-types"></a>
</h3>
<div class="callout-content">
<p>Workflow Inputs may be of different types, such as</p>
<ul>
<li><code>File</code></li>
<li><code>Directory</code></li>
<li><code>string</code></li>
<li><code>int</code></li>
<li><code>float</code></li>
<li><code>boolean</code></li>
<li><code>array</code></li>
<li><code>record</code></li>
<li><code>enum</code></li>
<li><code>Any</code></li>
<li>
<code>null</code>.</li>
</ul>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="workflow-steps">Workflow Steps<a class="anchor" aria-label="anchor" href="#workflow-steps"></a>
</h3>
<p>The next part of the script is the <code>steps</code> field.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">YAML<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode yaml" tabindex="0"><code class="sourceCode yaml"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="fu">steps</span><span class="kw">:</span></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a><span class="at">  </span><span class="fu">quality_control</span><span class="kw">:</span></span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span><span class="at"> bio-cwl-tools/fastqc/fastqc_2.cwl</span></span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a><span class="at">      </span><span class="fu">reads_file</span><span class="kw">:</span><span class="at"> rna_reads_fruitfly</span></span>
<span id="cb16-6"><a href="#cb16-6" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at">html_file</span><span class="kw">]</span></span></code></pre>
</div>
<p>Every step of a workflow needs a name, the first step of the workflow
is called <code>quality_control</code>.</p>
<p>Each step needs a <code>run</code> field, an <code>in</code> field
and an <code>out</code> field.</p>
<p>The <code>run</code> field contains the location of the CWL file of
the tool to be run OR the contents of the tool description itself.</p>
<p>The <code>in</code> field connects the <code>inputs</code> field to
the <code>fastqc</code> tool.</p>
<p>The <code>fastqc</code> tool has an input parameter called
<code>reads_file</code>, so it needs to connect the
<code>reads_file</code> to <code>rna_reads_fruitfly</code>.</p>
<p>Lastly, the <code>out</code> field is a list of output parameters
from the tool to be used.</p>
<p>In this example, the <code>fastqc</code> tool produces an output file
called <code>html_file</code>.</p>
</div>
<div class="section level3">
<h3 id="workflow-outputs">Workflow Outputs<a class="anchor" aria-label="anchor" href="#workflow-outputs"></a>
</h3>
<p>The last part of the script is the <code>output</code> field.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">YAML<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode yaml" tabindex="0"><code class="sourceCode yaml"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="fu">outputs</span><span class="kw">:</span></span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a><span class="at">  </span><span class="fu">quality_report</span><span class="kw">:</span></span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a><span class="at">    </span><span class="fu">outputSource</span><span class="kw">:</span><span class="at"> quality_control/html_file</span></span></code></pre>
</div>
<p>Each output in the <code>outputs</code> field needs its own name. In
this example the output is called <code>quality_report</code>.</p>
<p>Inside <code>quality_report</code> the type of output is defined. The
output of the <code>quality_control</code> step is a file, so the
<code>quality_report</code> type is <code>File</code>.</p>
<p>The <code>outputSource</code> field refers to where the output is
located after the commandline tool is complete, in this example it came
from the step <code>quality_control</code> and it is called
<code>html_file</code>.</p>
<div id="output-types" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="output-types" class="callout-inner">
<h3 class="callout-title">Output Types<a class="anchor" aria-label="anchor" href="#output-types"></a>
</h3>
<div class="callout-content">
<p>Like workflow inputs, workflow outputs can be of different types.
Refer to the list of input types above for the different types of
outputs for a workflow.</p>
</div>
</div>
</div>
</div>
</section><section id="running-the-workflow"><h2 class="section-heading">Running the workflow<a class="anchor" aria-label="anchor" href="#running-the-workflow"></a>
</h2>
<hr class="half-width">
<p>When you want to run this workflow, you need to provide a file with
the inputs the workflow needs. This file is similar to the
<code>hello_world.yml</code> file in the previous section. The input
file is called <code>workflow_input_1.yml</code></p>
<p><strong>workflow_input_1.yml</strong></p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">YAML<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode yaml" tabindex="0"><code class="sourceCode yaml"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="fu">rna_reads_fruitfly</span><span class="kw">:</span></span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a><span class="at">  </span><span class="fu">class</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a><span class="at">  </span><span class="fu">location</span><span class="kw">:</span><span class="at"> rnaseq/GSM461177_2_subsampled.fastqsanger</span></span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a><span class="at">  </span><span class="fu">format</span><span class="kw">:</span><span class="at"> http://edamontology.org/format_1930</span><span class="co">  # FASTA</span></span></code></pre>
</div>
<p>In the input file the values for the inputs that are declared in the
<code>inputs</code> section of the workflow are provided.</p>
<p>The workflow takes <code>rna_reads_fruitfly</code> as an input
parameter. We use the same variable name in the input file.</p>
<p>When using <code>File</code> or <code>Directory</code> inputs, the
class of the object needs to be defined, for example
<code>class: File</code> or <code>class: Directory</code>.</p>
<p>The <code>location</code> field contains the location of the input
file, in this case it is a local path, but we could have directly used
the original url
<code>location: https://zenodo.org/record/4541751/files/GSM461177_2_subsampled.fastqsanger</code></p>
<p>In this example the last line is needed to provide a format for the
fastq file. This is not always necessary, but it is good practice to
provide this information.</p>
<div id="specifying-the-location-of-an-input" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="specifying-the-location-of-an-input" class="callout-inner">
<h3 class="callout-title">Specifying the location of an input<a class="anchor" aria-label="anchor" href="#specifying-the-location-of-an-input"></a>
</h3>
<div class="callout-content">
<p>As shown above, we can use urls or local paths for the location of
the input file.</p>
<p>For local paths, if the path is a relative path, it should be
relative to the current working directory.</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="running-with-workflow">Running with workflow<a class="anchor" aria-label="anchor" href="#running-with-workflow"></a>
</h3>
<p>Now you can run the workflow using the following command:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="ex">cwltool</span> <span class="at">--cachedir</span> cache rna_seq_workflow_1.cwl workflow_input_1.yml</span></code></pre>
</div>
<div id="accordionSpoiler2" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler2" aria-expanded="false" aria-controls="collapseSpoiler2">
  <h3 class="accordion-header" id="headingSpoiler2">
<div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div>Console Outputs</h3>
</button>
<div id="collapseSpoiler2" class="accordion-collapse collapse" aria-labelledby="headingSpoiler2" data-bs-parent="#accordionSpoiler2">
<div class="accordion-body">
<pre><code>...
Analysis complete for GSM461177_2_subsampled.fastqsanger
INFO [job quality_control] Max memory used: 179MiB
INFO [job quality_control] completed success
INFO [step quality_control] completed success
INFO [workflow ] completed success
{
    "quality_report": {
        "location": "file:///.../GSM461177_2_subsampled.fastqsanger_fastqc.html",
        "basename": "GSM461177_2_subsampled.fastqsanger_fastqc.html",
        "class": "File",
        "checksum": "sha1$e820c530b91a3087ae4c53a6f9fbd35ab069095c",
        "size": 378324,
        "path": "/.../GSM461177_2_subsampled.fastqsanger_fastqc.html"
    }
}
INFO Final process status is success</code></pre>
</div>
</div>
</div>
</div>
<p><br></p>
<div id="cache-directory" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="cache-directory" class="callout-inner">
<h3 class="callout-title">Cache Directory<a class="anchor" aria-label="anchor" href="#cache-directory"></a>
</h3>
<div class="callout-content">
<p>To save intermediate results for re-use later we use
<code>--cachedir cache</code>; where <code>cache</code> is the directory
for storing the cache (it can be given any name, here we are just using
<code>cache</code> for simplicity). You can safely delete the
<code>cache</code> directory anytime, if you need to reclaim the disk
space.</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="challenges-1">Challenges<a class="anchor" aria-label="anchor" href="#challenges-1"></a>
</h3>
<div id="directly-embed-a-commandlinetool-into-a-file" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="directly-embed-a-commandlinetool-into-a-file" class="callout-inner">
<h3 class="callout-title">Directly embed a Commandlinetool into a file
üå∂üå∂<a class="anchor" aria-label="anchor" href="#directly-embed-a-commandlinetool-into-a-file"></a>
</h3>
<div class="callout-content">
<p>How could one embed the fastqc tool description directly into the
workflow?</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5">Show me the solution</h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" aria-labelledby="headingSolution5" data-bs-parent="#accordionSolution5">
<div class="accordion-body">
<p>The <code>run</code> field in the step can be replaced with the
<code>CommandLineTool</code> definition.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">YAML<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode yaml" tabindex="0"><code class="sourceCode yaml"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="fu">steps</span><span class="kw">:</span></span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a><span class="at">  </span><span class="fu">quality_control</span><span class="kw">:</span></span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span></span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a><span class="at">      </span><span class="fu">class</span><span class="kw">:</span><span class="at"> CommandLineTool</span></span>
<span id="cb21-5"><a href="#cb21-5" tabindex="-1"></a><span class="at">      </span><span class="fu">baseCommand</span><span class="kw">:</span><span class="at"> fastqc</span></span>
<span id="cb21-6"><a href="#cb21-6" tabindex="-1"></a><span class="at">      </span><span class="fu">inputs</span><span class="kw">:</span></span>
<span id="cb21-7"><a href="#cb21-7" tabindex="-1"></a><span class="at">        </span><span class="fu">reads_file</span><span class="kw">:</span></span>
<span id="cb21-8"><a href="#cb21-8" tabindex="-1"></a><span class="at">          </span><span class="fu">type</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb21-9"><a href="#cb21-9" tabindex="-1"></a><span class="at">          </span><span class="fu">inputBinding</span><span class="kw">:</span></span>
<span id="cb21-10"><a href="#cb21-10" tabindex="-1"></a><span class="at">            </span><span class="fu">position</span><span class="kw">:</span><span class="at"> </span><span class="dv">1</span></span>
<span id="cb21-11"><a href="#cb21-11" tabindex="-1"></a><span class="at">      </span><span class="fu">outputs</span><span class="kw">:</span></span>
<span id="cb21-12"><a href="#cb21-12" tabindex="-1"></a><span class="at">        </span><span class="fu">html_file</span><span class="kw">:</span></span>
<span id="cb21-13"><a href="#cb21-13" tabindex="-1"></a><span class="at">          </span><span class="fu">type</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb21-14"><a href="#cb21-14" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb21-15"><a href="#cb21-15" tabindex="-1"></a><span class="at">      </span><span class="fu">reads_file</span><span class="kw">:</span><span class="at"> rna_reads_fruitfly</span></span>
<span id="cb21-16"><a href="#cb21-16" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at">html_file</span><span class="kw">]</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="embedding-tool-descriptions" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="embedding-tool-descriptions" class="callout-inner">
<h3 class="callout-title">Embedding Tool Descriptions<a class="anchor" aria-label="anchor" href="#embedding-tool-descriptions"></a>
</h3>
<div class="callout-content">
<p>Embedding tool descriptions directly into the workflow can be useful
when the tool is very simple and only used once in the workflow.</p>
<p>This is not recommended for complex tools or tools that are used in
multiple workflows and as such is not common practise</p>
</div>
</div>
</div>
<div id="accordionInstructor1" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor1" aria-expanded="false" aria-controls="collapseInstructor1">
  <h3 class="accordion-header" id="headingInstructor1">
<div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div>Exercise</h3>
</button>
<div id="collapseInstructor1" class="accordion-collapse collapse" aria-labelledby="headingInstructor1" data-bs-parent="#accordionInstructor1">
<div class="accordion-body">
<p>Needs some exercises?</p>
<ul>
<li>Ask the students to get some information from each report generated
for the different data files we‚Äôve downloaded. This will involve them
making simple changes to the yaml configuration file.</li>
</ul>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>A tool description describes the interface to a command line
tool.</li>
<li>A workflow describes which command line tools to use in one or more
steps.</li>
<li>A tool descriptor is defined using the <code>ComandLineTool</code>
class.</li>
<li>How can we use a tool descriptor in a single step workflow?</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div>
</section></section><section id="aio-03-dependency_graphs"><p>Content from <a href="03-dependency_graphs.html">Developing Multi-Step Workflows</a></p>
<hr>
<p>Last updated on 2024-05-14 |
        
        <a href="https://github.com/__USERNAME__/__PROJECT__/edit/main/episodes/03-dependency_graphs.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 0 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can we expand to a multi-step workflow?</li>
<li>What is iterative workflow development?</li>
<li>How to use workflows as dependency graphs?</li>
<li>How to use sketches for workflow design?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain that a workflow is a dependency graph</li>
<li>Use cwlviewer online</li>
<li>Generate Graphviz diagram using cwltool</li>
<li>Exercise with the printout of a simple workflow; draw arrows on
code; hand draw a graph on another sheet of paper</li>
<li>Recognise that workflow development can be iterative i.e.¬†that it
doesn‚Äôt have to happen all at once</li>
<li>Understand the flow of data between tools</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="multi-step-workflow"><h2 class="section-heading">Multi-Step Workflow<a class="anchor" aria-label="anchor" href="#multi-step-workflow"></a>
</h2>
<hr class="half-width">
<p>In the previous episode we worked through a single step workflow,
carrying out quality control check on RNA reads of the fruitfly genome.
In this episode the workflow is extended with an equivalent reverse RNA
reads and the next two steps of the RNA-sequencing analysis, trimming
the reads and aligning the trimmed reads, are added. We will be using
the <a href="https://bio.tools/cutadapt" class="external-link"><code>cutadapt</code></a> and
<a href="https://bio.tools/star" class="external-link"><code>STAR</code></a> tools for these
tasks.</p>
<p>To make a multi-step workflow that can carry all this analysis out,
we add more entries to the <code>steps</code> field.</p>
<div id="naming-steps" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="naming-steps" class="callout-inner">
<h3 class="callout-title">Naming steps<a class="anchor" aria-label="anchor" href="#naming-steps"></a>
</h3>
<div class="callout-content">
<p>Note that when the <code>quality_control</code> step is duplicated
the two steps are named <code>quality_control_forward</code> and
<code>quality_control_reverse</code>, to distinguish the separate
forward and reverse RNA reads. Likewise, the
<code>rna_reads_fruitfly</code> input becomes
<code>rna_reads_fruitfly_forward</code>, and an
<code>rna_reads_fruitfly_reverse</code> input is added.</p>
</div>
</div>
</div>
<div id="accordionSpoiler1" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler1" aria-expanded="false" aria-controls="collapseSpoiler1">
  <h3 class="accordion-header" id="headingSpoiler1">
<div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div>rna_seq_workflow_2.cwl (Click to expand)</h3>
</button>
<div id="collapseSpoiler1" class="accordion-collapse collapse" data-bs-parent="#accordionSpoiler1" aria-labelledby="headingSpoiler1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">YAML<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode yaml" tabindex="0"><code class="sourceCode yaml"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">cwlVersion</span><span class="kw">:</span><span class="at"> v1.2</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">class</span><span class="kw">:</span><span class="at"> Workflow</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="fu">inputs</span><span class="kw">:</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="at">  </span><span class="fu">rna_reads_fruitfly_forward</span><span class="kw">:</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="at">    </span><span class="fu">format</span><span class="kw">:</span><span class="at"> http://edamontology.org/format_1930</span><span class="co">  # FASTQ</span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="at">  </span><span class="fu">rna_reads_fruitfly_reverse</span><span class="kw">:</span></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="at">    </span><span class="fu">format</span><span class="kw">:</span><span class="at"> http://edamontology.org/format_1930</span><span class="co">  # FASTQ</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a><span class="at">  </span><span class="fu">ref_fruitfly_genome</span><span class="kw">:</span><span class="at"> Directory</span></span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a><span class="at">  </span><span class="fu">fruitfly_gene_model</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a><span class="fu">steps</span><span class="kw">:</span></span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a><span class="at">  </span><span class="fu">quality_control_forward</span><span class="kw">:</span></span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span><span class="at"> bio-cwl-tools/fastqc/fastqc_2.cwl</span></span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a><span class="at">      </span><span class="fu">reads_file</span><span class="kw">:</span><span class="at"> rna_reads_fruitfly_forward</span></span>
<span id="cb1-19"><a href="#cb1-19" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at">html_file</span><span class="kw">]</span></span>
<span id="cb1-20"><a href="#cb1-20" tabindex="-1"></a></span>
<span id="cb1-21"><a href="#cb1-21" tabindex="-1"></a><span class="at">  </span><span class="fu">quality_control_reverse</span><span class="kw">:</span></span>
<span id="cb1-22"><a href="#cb1-22" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span><span class="at"> bio-cwl-tools/fastqc/fastqc_2.cwl</span></span>
<span id="cb1-23"><a href="#cb1-23" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb1-24"><a href="#cb1-24" tabindex="-1"></a><span class="at">      </span><span class="fu">reads_file</span><span class="kw">:</span><span class="at"> rna_reads_fruitfly_reverse</span></span>
<span id="cb1-25"><a href="#cb1-25" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at">html_file</span><span class="kw">]</span></span>
<span id="cb1-26"><a href="#cb1-26" tabindex="-1"></a></span>
<span id="cb1-27"><a href="#cb1-27" tabindex="-1"></a><span class="at">  </span><span class="fu">trim_low_quality_bases</span><span class="kw">:</span></span>
<span id="cb1-28"><a href="#cb1-28" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span><span class="at"> bio-cwl-tools/cutadapt/cutadapt-paired.cwl</span></span>
<span id="cb1-29"><a href="#cb1-29" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb1-30"><a href="#cb1-30" tabindex="-1"></a><span class="at">      </span><span class="fu">reads_1</span><span class="kw">:</span><span class="at"> rna_reads_fruitfly_forward</span></span>
<span id="cb1-31"><a href="#cb1-31" tabindex="-1"></a><span class="at">      </span><span class="fu">reads_2</span><span class="kw">:</span><span class="at"> rna_reads_fruitfly_reverse</span></span>
<span id="cb1-32"><a href="#cb1-32" tabindex="-1"></a><span class="at">      </span><span class="fu">minimum_length</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="at"> </span><span class="fu">default</span><span class="kw">:</span><span class="at"> </span><span class="dv">20</span><span class="at"> </span><span class="kw">}</span></span>
<span id="cb1-33"><a href="#cb1-33" tabindex="-1"></a><span class="at">      </span><span class="fu">quality_cutoff</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="at"> </span><span class="fu">default</span><span class="kw">:</span><span class="at"> </span><span class="dv">20</span><span class="at"> </span><span class="kw">}</span></span>
<span id="cb1-34"><a href="#cb1-34" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at"> trimmed_reads_</span><span class="dv">1</span><span class="kw">,</span><span class="at"> trimmed_reads_</span><span class="dv">2</span><span class="kw">,</span><span class="at"> report </span><span class="kw">]</span></span>
<span id="cb1-35"><a href="#cb1-35" tabindex="-1"></a></span>
<span id="cb1-36"><a href="#cb1-36" tabindex="-1"></a><span class="at">  </span><span class="fu">mapping_reads</span><span class="kw">:</span></span>
<span id="cb1-37"><a href="#cb1-37" tabindex="-1"></a><span class="at">    </span><span class="fu">requirements</span><span class="kw">:</span></span>
<span id="cb1-38"><a href="#cb1-38" tabindex="-1"></a><span class="at">      </span><span class="fu">ResourceRequirement</span><span class="kw">:</span></span>
<span id="cb1-39"><a href="#cb1-39" tabindex="-1"></a><span class="at">        </span><span class="fu">ramMin</span><span class="kw">:</span><span class="at"> </span><span class="dv">5120</span></span>
<span id="cb1-40"><a href="#cb1-40" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span><span class="at"> bio-cwl-tools/STAR/STAR-Align.cwl</span></span>
<span id="cb1-41"><a href="#cb1-41" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb1-42"><a href="#cb1-42" tabindex="-1"></a><span class="at">      </span><span class="fu">RunThreadN</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="fu">default</span><span class="kw">:</span><span class="at"> </span><span class="dv">4</span><span class="kw">}</span></span>
<span id="cb1-43"><a href="#cb1-43" tabindex="-1"></a><span class="at">      </span><span class="fu">GenomeDir</span><span class="kw">:</span><span class="at"> ref_fruitfly_genome</span></span>
<span id="cb1-44"><a href="#cb1-44" tabindex="-1"></a><span class="at">      </span><span class="fu">ForwardReads</span><span class="kw">:</span><span class="at"> trim_low_quality_bases/trimmed_reads_1</span></span>
<span id="cb1-45"><a href="#cb1-45" tabindex="-1"></a><span class="at">      </span><span class="fu">ReverseReads</span><span class="kw">:</span><span class="at"> trim_low_quality_bases/trimmed_reads_2</span></span>
<span id="cb1-46"><a href="#cb1-46" tabindex="-1"></a><span class="at">      </span><span class="fu">OutSAMtype</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="fu">default</span><span class="kw">:</span><span class="at"> BAM</span><span class="kw">}</span></span>
<span id="cb1-47"><a href="#cb1-47" tabindex="-1"></a><span class="at">      </span><span class="fu">SortedByCoordinate</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="fu">default</span><span class="kw">:</span><span class="at"> </span><span class="ch">true</span><span class="kw">}</span></span>
<span id="cb1-48"><a href="#cb1-48" tabindex="-1"></a><span class="at">      </span><span class="fu">OutSAMunmapped</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="fu">default</span><span class="kw">:</span><span class="at"> Within</span><span class="kw">}</span></span>
<span id="cb1-49"><a href="#cb1-49" tabindex="-1"></a><span class="at">      </span><span class="fu">Overhang</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="at"> </span><span class="fu">default</span><span class="kw">:</span><span class="at"> </span><span class="dv">36</span><span class="at"> </span><span class="kw">}</span><span class="co">  # the length of the reads - 1</span></span>
<span id="cb1-50"><a href="#cb1-50" tabindex="-1"></a><span class="at">      </span><span class="fu">Gtf</span><span class="kw">:</span><span class="at"> fruitfly_gene_model</span></span>
<span id="cb1-51"><a href="#cb1-51" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at">alignment</span><span class="kw">]</span></span>
<span id="cb1-52"><a href="#cb1-52" tabindex="-1"></a></span>
<span id="cb1-53"><a href="#cb1-53" tabindex="-1"></a><span class="at">  </span><span class="fu">index_alignment</span><span class="kw">:</span></span>
<span id="cb1-54"><a href="#cb1-54" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span><span class="at"> bio-cwl-tools/samtools/samtools_index.cwl</span></span>
<span id="cb1-55"><a href="#cb1-55" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb1-56"><a href="#cb1-56" tabindex="-1"></a><span class="at">      </span><span class="fu">bam_sorted</span><span class="kw">:</span><span class="at"> mapping_reads/alignment</span></span>
<span id="cb1-57"><a href="#cb1-57" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at">bam_sorted_indexed</span><span class="kw">]</span></span>
<span id="cb1-58"><a href="#cb1-58" tabindex="-1"></a></span>
<span id="cb1-59"><a href="#cb1-59" tabindex="-1"></a><span class="fu">outputs</span><span class="kw">:</span></span>
<span id="cb1-60"><a href="#cb1-60" tabindex="-1"></a><span class="at">  </span><span class="fu">quality_report_forward</span><span class="kw">:</span></span>
<span id="cb1-61"><a href="#cb1-61" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb1-62"><a href="#cb1-62" tabindex="-1"></a><span class="at">    </span><span class="fu">outputSource</span><span class="kw">:</span><span class="at"> quality_control_forward/html_file</span></span>
<span id="cb1-63"><a href="#cb1-63" tabindex="-1"></a><span class="at">  </span><span class="fu">quality_report_reverse</span><span class="kw">:</span></span>
<span id="cb1-64"><a href="#cb1-64" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb1-65"><a href="#cb1-65" tabindex="-1"></a><span class="at">    </span><span class="fu">outputSource</span><span class="kw">:</span><span class="at"> quality_control_reverse/html_file</span></span>
<span id="cb1-66"><a href="#cb1-66" tabindex="-1"></a><span class="at">  </span><span class="fu">bam_sorted_indexed</span><span class="kw">:</span></span>
<span id="cb1-67"><a href="#cb1-67" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb1-68"><a href="#cb1-68" tabindex="-1"></a><span class="at">    </span><span class="fu">outputSource</span><span class="kw">:</span><span class="at"> index_alignment/bam_sorted_indexed</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>The workflow file shows the first 5 steps of the RNA-seq analysis:
<code>quality_control_reverse</code>,
<code>quality_control_forward</code>,
<code>trim_low_quality_bases</code>, <code>mapping_reads</code>, and
<code>index_alignment</code>.</p>
<p>The <code>index_alignment</code> step uses the <code>alignment</code>
output of the <code>mapping_reads</code> step. You do this by
referencing the output of the <code>mapping_reads</code> step in the
<code>in</code> field of the <code>index_alignment</code> step. This is
similar to referencing the outputs of the different steps in the
<code>outputs</code> section.</p>
<div id="default-values" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="default-values" class="callout-inner">
<h3 class="callout-title">Default values<a class="anchor" aria-label="anchor" href="#default-values"></a>
</h3>
<div class="callout-content">
<p>The <code>mapping_reads</code> step needs some extra information
beyond the inputs from the other steps, which is done by providing
<code>default</code> values. If you want, you can read the
<code>bio-cwl-tools/STAR/STAR-Align.cwl</code> file to see how these
extra inputs are transformed into command line options to the
<code>STAR</code> program. This information is provided in the
<code>in</code> field.</p>
</div>
</div>
</div>
<div id="specifying-resources" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="specifying-resources" class="callout-inner">
<h3 class="callout-title">Specifying resources<a class="anchor" aria-label="anchor" href="#specifying-resources"></a>
</h3>
<div class="callout-content">
<p>To run the tool better, it needs more RAM than the default. So there
is a <code>requirements</code> entry inside the
<code>mapping_reads</code> step definition with a
<code>ResourceRequirement</code> to allocate a minimum of 5120 MiB (5
GiB) of RAM.</p>
</div>
</div>
</div>
<p>The newly added <code>mapping_reads</code> step also need an input
not provided by any of our other steps, therefore an additional
workflow-level input is added: a directory that contains the reference
genome necessary for the mapping.</p>
<p>This <code>ref_fruitfly_genome</code> is added in the
<code>inputs</code> field of the workflow and in the YAML input file,
<code>workflow_input_2.yml</code>.</p>
</section><section id="validating-the-workflow"><h2 class="section-heading">Validating the workflow<a class="anchor" aria-label="anchor" href="#validating-the-workflow"></a>
</h2>
<hr class="half-width">
<div id="is-this-a-valid-workflow" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="is-this-a-valid-workflow" class="callout-inner">
<h3 class="callout-title">Is this a valid workflow? üå∂<a class="anchor" aria-label="anchor" href="#is-this-a-valid-workflow"></a>
</h3>
<div class="callout-content">
<p>Use cwltool to validate the workflow</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">Show me the solution</h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">cwltool</span> <span class="at">--validate</span> rna_seq_workflow_2.cwl</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>A warning is thrown after we validate this workflow.</p>
<pre><code>WARNING Workflow checker warning:
rna_seq_workflow_2.cwl:51:11: Source 'alignment' of type ["File", {"type": "array", "items":
                              "File"}] may be incompatible
rna_seq_workflow_2.cwl:56:7:    with sink 'bam_sorted' of type "File"</code></pre>
<p><br></p>
<div id="should-we-be-concerned-about-this-warning" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="should-we-be-concerned-about-this-warning" class="callout-inner">
<h3 class="callout-title">Should we be concerned about this warning
üå∂üå∂üå∂<a class="anchor" aria-label="anchor" href="#should-we-be-concerned-about-this-warning"></a>
</h3>
<div class="callout-content">
<p>Not all warnings are bad.<br>
Should we be concerned about this warning? If not, why not?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">Show me the solution</h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>In this case, the outputs of the mapping step may instead either be
just one file, OR an array of files.</p>
<p>It is important that the invoke the <code>mapping</code> step in such
a way that only one file is output, as a single file is the requirement
of the <code>index_alignment</code> step</p>
</div>
</div>
</div>
</div>
</section><section id="running-the-new-workflow"><h2 class="section-heading">Running the new workflow<a class="anchor" aria-label="anchor" href="#running-the-new-workflow"></a>
</h2>
<hr class="half-width">
<p>The workflow definition is complete and we now only need to write the
YAML input file.</p>
<div id="accordionSpoiler2" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler2" aria-expanded="false" aria-controls="collapseSpoiler2">
  <h3 class="accordion-header" id="headingSpoiler2">
<div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div>workflow_input_2.yml (Click to expand)</h3>
</button>
<div id="collapseSpoiler2" class="accordion-collapse collapse" data-bs-parent="#accordionSpoiler2" aria-labelledby="headingSpoiler2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">YAML<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode yaml" tabindex="0"><code class="sourceCode yaml"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">rna_reads_fruitfly_forward</span><span class="kw">:</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="at">  </span><span class="fu">class</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="at">  </span><span class="fu">location</span><span class="kw">:</span><span class="at"> rnaseq/GSM461177_1_subsampled.fastqsanger</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="at">  </span><span class="fu">format</span><span class="kw">:</span><span class="at"> http://edamontology.org/format_1930</span><span class="co">  # FASTQ</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="fu">rna_reads_fruitfly_reverse</span><span class="kw">:</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="at">  </span><span class="fu">class</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="at">  </span><span class="fu">location</span><span class="kw">:</span><span class="at"> rnaseq/GSM461177_2_subsampled.fastqsanger</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="at">  </span><span class="fu">format</span><span class="kw">:</span><span class="at"> http://edamontology.org/format_1930</span><span class="co">  # FASTQ</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a><span class="fu">ref_fruitfly_genome</span><span class="kw">:</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a><span class="at">  </span><span class="fu">class</span><span class="kw">:</span><span class="at"> Directory</span></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a><span class="at">  </span><span class="fu">location</span><span class="kw">:</span><span class="at"> rnaseq/dm6-STAR-index</span></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a><span class="fu">fruitfly_gene_model</span><span class="kw">:</span></span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a><span class="at">  </span><span class="fu">class</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a><span class="at">  </span><span class="fu">location</span><span class="kw">:</span><span class="at"> rnaseq/Drosophila_melanogaster.BDGP6.87.gtf</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>We have finished the workflow definition and the input file and now
can run the workflow.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">cwltool</span> <span class="at">--cachedir</span> cache rna_seq_workflow_2.cwl workflow_input_2.yml</span></code></pre>
</div>
<div id="challenge-draw-the-workflow" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-draw-the-workflow" class="callout-inner">
<h3 class="callout-title">Challenge: Draw the workflow üå∂<a class="anchor" aria-label="anchor" href="#challenge-draw-the-workflow"></a>
</h3>
<div class="callout-content">
<p>Draw the connecting arrows in the following graph of our workflow.
Also, provide the outputs/inputs of the different steps. You can use for
example Paint or print out the graph.</p>
<figure><img src="../fig/Ep3_empty_graph.png" style='height: "300px"' alt="Ep3 Empty Graph" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">Show me the solution</h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<p>To find out how the inputs and the steps are connected to each other,
look at the <code>in</code> field of the different steps.</p>
<figure><img src="../fig/Ep3_graph_answer.png" style='height: "300px"' alt="Ep3 empty graph answer" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
<div id="iterative-working" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="iterative-working" class="callout-inner">
<h3 class="callout-title">Iterative working<a class="anchor" aria-label="anchor" href="#iterative-working"></a>
</h3>
<div class="callout-content">
<p>Working on a workflow is often not something that happens all at
once. Sometimes you already have a shell script ready that can be
converted to a CWL workflow. Other times it is similar to this tutorial,
you start with a single-step workflow and extend it to a multi-step
workflow. This is all iterative working, a continuous work in
progress.</p>
</div>
</div>
</div>
</section><section id="visualising-a-workflow"><h2 class="section-heading">Visualising a workflow<a class="anchor" aria-label="anchor" href="#visualising-a-workflow"></a>
</h2>
<hr class="half-width">
<p>To visualise a workflow, a graph can be used. This can be done before
a CWL script is written to visualise how the different steps connect to
each other. It is also possible to make a graph after the CWL script has
been written. This graph can be generated using online tools or the
built-in function in <code>cwltool</code>. When a graph is generated, it
can be used to visualise the steps taken and could make it easier to
explain a workflow to other researchers.</p>
<p>A CWL workflow is a directed acyclic graph (DAG). This means
that:</p>
<ol style="list-style-type: decimal">
<li>The workflow has a certain direction, from workflow inputs to step
inputs, from step outputs to other step inputs, and from step outputs to
workflow outputs and</li>
<li>The workflow definition has no cycles.</li>
</ol>
<div id="cwl-workflow-as-a-dependency-graph" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="cwl-workflow-as-a-dependency-graph" class="callout-inner">
<h3 class="callout-title">CWL workflow as a dependency graph<a class="anchor" aria-label="anchor" href="#cwl-workflow-as-a-dependency-graph"></a>
</h3>
<div class="callout-content">
<p>By being a ‚ÄòDAG‚Äô, a CWL workflow is a dependency graph. Each input
for a step in the workflow depends on either a workflow-level input or
the presence of a particular output from another step.</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="from-cwl-script-to-graph">From CWL script to graph<a class="anchor" aria-label="anchor" href="#from-cwl-script-to-graph"></a>
</h3>
<p>In this example the workflow is already made, so the graph can be
generated using <a href="https://view.commonwl.org/" class="external-link">cwlviewer</a>
online or using <code>cwltool</code>. First, let‚Äôs have a look at <a href="https://view.commonwl.org/" class="external-link">cwlviewer</a>. To use this tool, the
workflow has to be put in a GitHub, GitLab or Git repository. To view
the graph of the workflow enter the URL and click
<code>Parse Workflow</code>.</p>
<div id="push-your-workflow-to-github" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="push-your-workflow-to-github" class="callout-inner">
<h3 class="callout-title">Push your workflow to GitHub üå∂<a class="anchor" aria-label="anchor" href="#push-your-workflow-to-github"></a>
</h3>
<div class="callout-content">
<p>Add your workflow to a git commit and then push that commit to
github.com</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4">Git add / commit / push</h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<p>Your solution might look like this</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">git</span> add rna_seq_workflow_2.cwl </span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="fu">git</span> commit <span class="at">-m</span> <span class="st">"Added my second RNASeq CWL Workflow"</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="fu">git</span> push</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Now it‚Äôs time to view your workflow!</p>
<div id="view-your-workflow-in-the-cwl-viewer" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="view-your-workflow-in-the-cwl-viewer" class="callout-inner">
<h3 class="callout-title">View your workflow in the cwl viewer üå∂<a class="anchor" aria-label="anchor" href="#view-your-workflow-in-the-cwl-viewer"></a>
</h3>
<div class="callout-content">
<p>Paste the workflow url into the form on <a href="https://view.commonwl.org/" class="external-link">view.commonwl.org</a></p>
<p>Your workflow url will be something like
<code>https://github.com/alexiswl/cwl-novice-tutorial/blob/main/rna_seq_workflow_2.cwl</code>.</p>
</div>
</div>
</div>
<p>The cwlviewer displays the workflow as a graph, starting with the
input. Then the different steps are shown, each with their input(s) and
output(s). The steps are linked to each other using arrows accompanied
by the input of the next step. The graph ends with the workflow
outputs.</p>
<p>The graph of the RNA-seq workflow looks a follows:</p>
<figure><img src="../fig/Ep3_graph_answer.png" style='height: "400px"' alt="Ep3 graph answer" class="figure mx-auto d-block"></figure>
</div>
<div class="section level3">
<h3 id="generating-graphs-locally">Generating graphs locally<a class="anchor" aria-label="anchor" href="#generating-graphs-locally"></a>
</h3>
<p>It is also possible to generate the graph in the command line.
<code>cwltool</code> has a function that makes a graph. The
<code>--print-dot</code> option will print a file suitable for Graphviz
<code>dot</code> program. This is the command to generate a Scalable
Vector Graphic (SVG) file:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="ex">cwltool</span> <span class="at">--print-dot</span> rna_seq_workflow_2.cwl <span class="kw">|</span> <span class="ex">dot</span> <span class="at">-Tsvg</span> <span class="op">&gt;</span> workflow_graph_2.svg</span></code></pre>
</div>
<p>The resulting SVG file displays the same graph as the one in the
cwlviewer. The SVG file can be opened in any web browser and in <a href="https://inkscape.org/" class="external-link">Inkscape</a>, for example. Or opened with
<code>code workflow_graph_2.svg</code> from the terminal.</p>
<div id="windows-only-view-images-from-the-cli-with-wslview" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="windows-only-view-images-from-the-cli-with-wslview" class="callout-inner">
<h3 class="callout-title">Windows Only: View images from the CLI with
wslview<a class="anchor" aria-label="anchor" href="#windows-only-view-images-from-the-cli-with-wslview"></a>
</h3>
<div class="callout-content">
<p>Windows users can run <code>wslview workflow_graph_2.svg</code> in
their terminal to view the graph in the default web browser.</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="visualisation-in-vscode">Visualisation in VSCode<a class="anchor" aria-label="anchor" href="#visualisation-in-vscode"></a>
</h3>
<p><a href="https://marketplace.visualstudio.com/items?itemName=sbg-rabix.benten-cwl" class="external-link"><strong>Benten</strong></a>
is an extension in Visual Studio Code (VSCode) that among other things
visualises a workflow in a graph. When Benten is installed in VSCode,
the tool can be used to visualise the workflow. In the top-right corner
of the VSCode window the CWL viewer can be opened, see the screenshot
below.</p>
<figure><img src="../fig/VSCode_CWL_Preview_(step1).png" style='height: "400px"' alt="VSCode CWL Preview step 1" class="figure mx-auto d-block"></figure><p>In VSCode/Benten the inputs are shown in green, the steps in blue and
the outputs in yellow. This graph looks a little bit different from the
graph made with cwlviewer or <code>cwltool</code>. The graph by
VSCode/Benten doesn‚Äôt show the output-input names between the different
steps.</p>
<figure><img src="../fig/VSCode_CWL_Preview_(step2).png" style='height: "400px"' alt="VSCode CWL Preview step 2" class="figure mx-auto d-block"></figure><div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>A multi-step workflow has multiple entries under the
<code>steps</code> section</li>
<li>Workflow development can be an iterative process</li>
<li>A CWL workflow can be represented as a dependency graph, either to
explain your workflow or as a planning tool</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div>
</section></section><section id="aio-04-reusing_tools"><p>Content from <a href="04-reusing_tools.html">Resources for Reusing Tools and Scripts</a></p>
<hr>
<p>Last updated on 2024-05-14 |
        
        <a href="https://github.com/__USERNAME__/__PROJECT__/edit/main/episodes/04-reusing_tools.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 0 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How to find other solutions/CWL recipes for awkward problems?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Know good resources for finding solutions to common problems</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="pre-written-tool-descriptions"><h2 class="section-heading">Pre-written tool descriptions<a class="anchor" aria-label="anchor" href="#pre-written-tool-descriptions"></a>
</h2>
<hr class="half-width">
<p>When you start a CWL workflow, it is recommended to check if there is
already a CWL document available for the tools you want to use. <a href="https://github.com/common-workflow-library/bio-cwl-tools" class="external-link">Bio-cwl-tools</a>
is a library of CWL documents for biology/life-sciences related
tools.</p>
<p>The CWL documents of the previous steps were already provided for
you, however, you can also find them in this library. In this episode
you will use the bio-cwl-tools library to add the last step to the
workflow.</p>
</section><section id="adding-new-step-in-workflow"><h2 class="section-heading">Adding new step in workflow<a class="anchor" aria-label="anchor" href="#adding-new-step-in-workflow"></a>
</h2>
<hr class="half-width">
<p>The last step of our workflow is counting the RNA-seq reads for which
we will use the <a href="https://bio.tools/featurecounts" class="external-link"><code>featureCounts</code></a>
tool.</p>
<div id="find-the-featurecounts-tool-in-the-bio-cwl-tools-library" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="find-the-featurecounts-tool-in-the-bio-cwl-tools-library" class="callout-inner">
<h3 class="callout-title">Find the featureCounts tool in the
bio-cwl-tools library üå∂<a class="anchor" aria-label="anchor" href="#find-the-featurecounts-tool-in-the-bio-cwl-tools-library"></a>
</h3>
<div class="callout-content">
<p>Find the <code>featureCounts</code> tool in the <a href="https://github.com/common-workflow-library/bio-cwl-tools" class="external-link">bio-cwl-tools
library</a>. Have a look at the CWL document. Which inputs does this
tool need? And what are the outputs of this tool?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">Show me the solution</h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>The <code>featureCounts</code> CWL document can be found in the <a href="https://github.com/common-workflow-library/bio-cwl-tools/blob/release/subread/featureCounts.cwl" class="external-link">GitHub
repo</a>.</p>
<p>It has three inputs: - annotations - A GTF or GFF file containing the
gene annotations - mapped_reads - A BAM file containing the mapped reads
- reads_are_paired - A boolean value. These inputs can be found on lines
6, 9, and 12.</p>
<p>The output of this tool is a file called <code>featurecounts</code>
(line 27).</p>
</div>
</div>
</div>
</div>
<p><br></p>
<div class="section level3">
<h3 id="appending-the-featurecounts-step-to-the-workflow">Appending the featureCounts step to the workflow<a class="anchor" aria-label="anchor" href="#appending-the-featurecounts-step-to-the-workflow"></a>
</h3>
<p>We need a local copy of <code>featureCounts</code> in order to use it
in our workflow.</p>
<p>We already imported this as a git submodule during setup, so the tool
should be located at
<code>bio-cwl-tools/subread/featureCounts.cwl</code>.</p>
<div id="add-the-featurecounts-tool-to-the-workflow" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="add-the-featurecounts-tool-to-the-workflow" class="callout-inner">
<h3 class="callout-title">Add the featureCounts tool to the workflow
üå∂üå∂<a class="anchor" aria-label="anchor" href="#add-the-featurecounts-tool-to-the-workflow"></a>
</h3>
<div class="callout-content">
<p>Please copy the <code>rna_seq_workflow_2.cwl</code> file to create
<code>rna_seq_workflow_3.cwl</code>.</p>
<p>Add the <code>featureCounts</code> tool to the workflow as a workflow
step.</p>
<p><strong>Bonus</strong>: üå∂üå∂üå∂</p>
<p>Similar to the <code>STAR</code> tool, this tool also needs more RAM
than the default.</p>
<p>Update the RAM used to run the tool without editing the
commandlinetool</p>
</div>
</div>
</div>
<div id="accordionHint1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint1" aria-expanded="false" aria-controls="collapseHint1">
  <h4 class="accordion-header" id="headingHint1">Hint: Feature Counts Workflow Step</h4>
</button>
<div id="collapseHint1" class="accordion-collapse collapse" data-bs-parent="#accordionHint1" aria-labelledby="headingHint1">
<div class="accordion-body">
<p>Use the <code>run</code> field to add the <code>featureCounts</code>
tool as a step in the workflow.</p>
</div>
</div>
</div>
</div>
<div id="accordionHint2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint2" aria-expanded="false" aria-controls="collapseHint2">
  <h4 class="accordion-header" id="headingHint2">Hint: Upgrade RAM</h4>
</button>
<div id="collapseHint2" class="accordion-collapse collapse" data-bs-parent="#accordionHint2" aria-labelledby="headingHint2">
<div class="accordion-body">
<p>Use a <code>requirements</code> entry with
<code>ResourceRequirement</code> to allocate a <code>ramMin</code> of
500.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">Solution: workflow_3.yml (Click to expand)</h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">YAML<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode yaml" tabindex="0"><code class="sourceCode yaml"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">cwlVersion</span><span class="kw">:</span><span class="at"> v1.2</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">class</span><span class="kw">:</span><span class="at"> Workflow</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">inputs</span><span class="kw">:</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="at">  </span><span class="fu">rna_reads_fruitfly_forward</span><span class="kw">:</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="at">    </span><span class="fu">format</span><span class="kw">:</span><span class="at"> http://edamontology.org/format_1930</span><span class="co">  # FASTQ</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="at">  </span><span class="fu">rna_reads_fruitfly_reverse</span><span class="kw">:</span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="at">    </span><span class="fu">format</span><span class="kw">:</span><span class="at"> http://edamontology.org/format_1930</span><span class="co">  # FASTQ</span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="at">  </span><span class="fu">ref_fruitfly_genome</span><span class="kw">:</span><span class="at"> Directory</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a><span class="at">  </span><span class="fu">fruitfly_gene_model</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a><span class="fu">steps</span><span class="kw">:</span></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a><span class="at">  </span><span class="fu">quality_control_forward</span><span class="kw">:</span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span><span class="at"> bio-cwl-tools/fastqc/fastqc_2.cwl</span></span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a><span class="at">      </span><span class="fu">reads_file</span><span class="kw">:</span><span class="at"> rna_reads_fruitfly_forward</span></span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at">html_file</span><span class="kw">]</span></span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a><span class="at">  </span><span class="fu">quality_control_reverse</span><span class="kw">:</span></span>
<span id="cb1-19"><a href="#cb1-19" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span><span class="at"> bio-cwl-tools/fastqc/fastqc_2.cwl</span></span>
<span id="cb1-20"><a href="#cb1-20" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb1-21"><a href="#cb1-21" tabindex="-1"></a><span class="at">      </span><span class="fu">reads_file</span><span class="kw">:</span><span class="at"> rna_reads_fruitfly_reverse</span></span>
<span id="cb1-22"><a href="#cb1-22" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at">html_file</span><span class="kw">]</span></span>
<span id="cb1-23"><a href="#cb1-23" tabindex="-1"></a><span class="at">  </span><span class="fu">trim_low_quality_bases</span><span class="kw">:</span></span>
<span id="cb1-24"><a href="#cb1-24" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span><span class="at"> bio-cwl-tools/cutadapt/cutadapt-paired.cwl</span></span>
<span id="cb1-25"><a href="#cb1-25" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb1-26"><a href="#cb1-26" tabindex="-1"></a><span class="at">      </span><span class="fu">reads_1</span><span class="kw">:</span><span class="at"> rna_reads_fruitfly_forward</span></span>
<span id="cb1-27"><a href="#cb1-27" tabindex="-1"></a><span class="at">      </span><span class="fu">reads_2</span><span class="kw">:</span><span class="at"> rna_reads_fruitfly_reverse</span></span>
<span id="cb1-28"><a href="#cb1-28" tabindex="-1"></a><span class="at">      </span><span class="fu">minimum_length</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="at"> </span><span class="fu">default</span><span class="kw">:</span><span class="at"> </span><span class="dv">20</span><span class="at"> </span><span class="kw">}</span></span>
<span id="cb1-29"><a href="#cb1-29" tabindex="-1"></a><span class="at">      </span><span class="fu">quality_cutoff</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="at"> </span><span class="fu">default</span><span class="kw">:</span><span class="at"> </span><span class="dv">20</span><span class="at"> </span><span class="kw">}</span></span>
<span id="cb1-30"><a href="#cb1-30" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at"> trimmed_reads_</span><span class="dv">1</span><span class="kw">,</span><span class="at"> trimmed_reads_</span><span class="dv">2</span><span class="kw">,</span><span class="at"> report </span><span class="kw">]</span></span>
<span id="cb1-31"><a href="#cb1-31" tabindex="-1"></a><span class="at">  </span><span class="fu">mapping_reads</span><span class="kw">:</span></span>
<span id="cb1-32"><a href="#cb1-32" tabindex="-1"></a><span class="at">    </span><span class="fu">requirements</span><span class="kw">:</span></span>
<span id="cb1-33"><a href="#cb1-33" tabindex="-1"></a><span class="at">      </span><span class="fu">ResourceRequirement</span><span class="kw">:</span></span>
<span id="cb1-34"><a href="#cb1-34" tabindex="-1"></a><span class="at">        </span><span class="fu">ramMin</span><span class="kw">:</span><span class="at"> </span><span class="dv">5120</span></span>
<span id="cb1-35"><a href="#cb1-35" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span><span class="at"> bio-cwl-tools/STAR/STAR-Align.cwl</span></span>
<span id="cb1-36"><a href="#cb1-36" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb1-37"><a href="#cb1-37" tabindex="-1"></a><span class="at">      </span><span class="fu">RunThreadN</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="fu">default</span><span class="kw">:</span><span class="at"> </span><span class="dv">4</span><span class="kw">}</span></span>
<span id="cb1-38"><a href="#cb1-38" tabindex="-1"></a><span class="at">      </span><span class="fu">GenomeDir</span><span class="kw">:</span><span class="at"> ref_fruitfly_genome</span></span>
<span id="cb1-39"><a href="#cb1-39" tabindex="-1"></a><span class="at">      </span><span class="fu">ForwardReads</span><span class="kw">:</span><span class="at"> trim_low_quality_bases/trimmed_reads_1</span></span>
<span id="cb1-40"><a href="#cb1-40" tabindex="-1"></a><span class="at">      </span><span class="fu">ReverseReads</span><span class="kw">:</span><span class="at"> trim_low_quality_bases/trimmed_reads_2</span></span>
<span id="cb1-41"><a href="#cb1-41" tabindex="-1"></a><span class="at">      </span><span class="fu">OutSAMtype</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="fu">default</span><span class="kw">:</span><span class="at"> BAM</span><span class="kw">}</span></span>
<span id="cb1-42"><a href="#cb1-42" tabindex="-1"></a><span class="at">      </span><span class="fu">SortedByCoordinate</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="fu">default</span><span class="kw">:</span><span class="at"> </span><span class="ch">true</span><span class="kw">}</span></span>
<span id="cb1-43"><a href="#cb1-43" tabindex="-1"></a><span class="at">      </span><span class="fu">OutSAMunmapped</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="fu">default</span><span class="kw">:</span><span class="at"> Within</span><span class="kw">}</span></span>
<span id="cb1-44"><a href="#cb1-44" tabindex="-1"></a><span class="at">      </span><span class="fu">Overhang</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="at"> </span><span class="fu">default</span><span class="kw">:</span><span class="at"> </span><span class="dv">36</span><span class="at"> </span><span class="kw">}</span><span class="co">  # the length of the reads - 1</span></span>
<span id="cb1-45"><a href="#cb1-45" tabindex="-1"></a><span class="at">      </span><span class="fu">Gtf</span><span class="kw">:</span><span class="at"> fruitfly_gene_model</span></span>
<span id="cb1-46"><a href="#cb1-46" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at">alignment</span><span class="kw">]</span></span>
<span id="cb1-47"><a href="#cb1-47" tabindex="-1"></a><span class="at">  </span><span class="fu">index_alignment</span><span class="kw">:</span></span>
<span id="cb1-48"><a href="#cb1-48" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span><span class="at"> bio-cwl-tools/samtools/samtools_index.cwl</span></span>
<span id="cb1-49"><a href="#cb1-49" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb1-50"><a href="#cb1-50" tabindex="-1"></a><span class="at">      </span><span class="fu">bam_sorted</span><span class="kw">:</span><span class="at"> mapping_reads/alignment</span></span>
<span id="cb1-51"><a href="#cb1-51" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at">bam_sorted_indexed</span><span class="kw">]</span></span>
<span id="cb1-52"><a href="#cb1-52" tabindex="-1"></a><span class="at">  </span><span class="fu">count_reads</span><span class="kw">:</span></span>
<span id="cb1-53"><a href="#cb1-53" tabindex="-1"></a><span class="at">    </span><span class="fu">requirements</span><span class="kw">:</span></span>
<span id="cb1-54"><a href="#cb1-54" tabindex="-1"></a><span class="at">      </span><span class="fu">ResourceRequirement</span><span class="kw">:</span></span>
<span id="cb1-55"><a href="#cb1-55" tabindex="-1"></a><span class="at">        </span><span class="fu">ramMin</span><span class="kw">:</span><span class="at"> </span><span class="dv">500</span></span>
<span id="cb1-56"><a href="#cb1-56" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span><span class="at"> bio-cwl-tools/subread/featureCounts.cwl</span></span>
<span id="cb1-57"><a href="#cb1-57" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb1-58"><a href="#cb1-58" tabindex="-1"></a><span class="at">      </span><span class="fu">mapped_reads</span><span class="kw">:</span><span class="at"> index_alignment/bam_sorted_indexed</span></span>
<span id="cb1-59"><a href="#cb1-59" tabindex="-1"></a><span class="at">      </span><span class="fu">annotations</span><span class="kw">:</span><span class="at"> fruitfly_gene_model</span></span>
<span id="cb1-60"><a href="#cb1-60" tabindex="-1"></a><span class="at">      </span><span class="fu">reads_are_paired</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="fu">default</span><span class="kw">:</span><span class="at"> </span><span class="ch">true</span><span class="kw">}</span></span>
<span id="cb1-61"><a href="#cb1-61" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at">featurecounts</span><span class="kw">]</span></span>
<span id="cb1-62"><a href="#cb1-62" tabindex="-1"></a><span class="fu">outputs</span><span class="kw">:</span></span>
<span id="cb1-63"><a href="#cb1-63" tabindex="-1"></a><span class="at">  </span><span class="fu">quality_report_forward</span><span class="kw">:</span></span>
<span id="cb1-64"><a href="#cb1-64" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb1-65"><a href="#cb1-65" tabindex="-1"></a><span class="at">    </span><span class="fu">outputSource</span><span class="kw">:</span><span class="at"> quality_control_forward/html_file</span></span>
<span id="cb1-66"><a href="#cb1-66" tabindex="-1"></a><span class="at">  </span><span class="fu">quality_report_reverse</span><span class="kw">:</span></span>
<span id="cb1-67"><a href="#cb1-67" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb1-68"><a href="#cb1-68" tabindex="-1"></a><span class="at">    </span><span class="fu">outputSource</span><span class="kw">:</span><span class="at"> quality_control_reverse/html_file</span></span>
<span id="cb1-69"><a href="#cb1-69" tabindex="-1"></a><span class="at">  </span><span class="fu">bam_sorted_indexed</span><span class="kw">:</span></span>
<span id="cb1-70"><a href="#cb1-70" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb1-71"><a href="#cb1-71" tabindex="-1"></a><span class="at">    </span><span class="fu">outputSource</span><span class="kw">:</span><span class="at"> index_alignment/bam_sorted_indexed</span></span>
<span id="cb1-72"><a href="#cb1-72" tabindex="-1"></a><span class="at">  </span><span class="fu">featurecounts</span><span class="kw">:</span></span>
<span id="cb1-73"><a href="#cb1-73" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb1-74"><a href="#cb1-74" tabindex="-1"></a><span class="at">    </span><span class="fu">outputSource</span><span class="kw">:</span><span class="at"> count_reads/featurecounts</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section id="running-the-new-workflow"><h2 class="section-heading">Running the new workflow<a class="anchor" aria-label="anchor" href="#running-the-new-workflow"></a>
</h2>
<hr class="half-width">
<p>The workflow is complete and we only need to complete the YAML input
file.</p>
<p>Copy the <code>workflow_input_2.yml</code> file to
<code>workflow_input_3.yml</code>, and add the last entry in the input
file, which is the <code>fruitfly_gene_model</code> file.</p>
<div id="accordionSpoiler1" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler1" aria-expanded="false" aria-controls="collapseSpoiler1">
  <h3 class="accordion-header" id="headingSpoiler1">
<div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div>workflow_input_3.yml (Click to expand)</h3>
</button>
<div id="collapseSpoiler1" class="accordion-collapse collapse" data-bs-parent="#accordionSpoiler1" aria-labelledby="headingSpoiler1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">YAML<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode yaml" tabindex="0"><code class="sourceCode yaml"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">rna_reads_fruitfly_forward</span><span class="kw">:</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="at">  </span><span class="fu">class</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="at">  </span><span class="fu">location</span><span class="kw">:</span><span class="at"> rnaseq/GSM461177_1_subsampled.fastqsanger</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="at">  </span><span class="fu">format</span><span class="kw">:</span><span class="at"> http://edamontology.org/format_1930</span><span class="co">  # FASTQ</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="fu">rna_reads_fruitfly_reverse</span><span class="kw">:</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="at">  </span><span class="fu">class</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="at">  </span><span class="fu">location</span><span class="kw">:</span><span class="at"> rnaseq/GSM461177_2_subsampled.fastqsanger</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="at">  </span><span class="fu">format</span><span class="kw">:</span><span class="at"> http://edamontology.org/format_1930</span><span class="co">  # FASTQ</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="fu">ref_fruitfly_genome</span><span class="kw">:</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="at">  </span><span class="fu">class</span><span class="kw">:</span><span class="at"> Directory</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="at">  </span><span class="fu">location</span><span class="kw">:</span><span class="at"> rnaseq/dm6-STAR-index</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="fu">fruitfly_gene_model</span><span class="kw">:</span></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a><span class="at">  </span><span class="fu">class</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a><span class="at">  </span><span class="fu">location</span><span class="kw">:</span><span class="at"> rnaseq/Drosophila_melanogaster.BDGP6.87.gtf</span></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a><span class="at">  </span><span class="fu">format</span><span class="kw">:</span><span class="at"> http://edamontology.org/format_2306</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="prereq1" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Prerequisite<a class="anchor" aria-label="anchor" href="#prereq1"></a>
</h3>
<div class="callout-content">
<p>You have finished the workflow and the input file and now you can run
the whole workflow.</p>
</div>
</div>
</div>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">cwltool</span> <span class="at">--cachedir</span> cache rna_seq_workflow_3.cwl workflow_input_3.yml</span></code></pre>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>
<a href="https://github.com/common-workflow-library/bio-cwl-tools" class="external-link">bio-cwl-tools</a>
is a library of CWL documents for biology/life-sciences related
tools</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></section><section id="aio-addendum-01-debug"><p>Content from <a href="addendum-01-debug.html">Debugging Workflows</a></p>
<hr>
<p>Last updated on 2024-05-14 |
        
        <a href="https://github.com/__USERNAME__/__PROJECT__/edit/main/episodes/addendum-01-debug.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 0 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li><p>‚ÄúHow can I check my CWL file for errors?‚Äù</p></li>
<li><p>‚ÄúHow can I get more information to help with solving an
error?‚Äù</p></li>
<li><p>‚ÄúWhat are some common error messages when using CWL?‚Äù</p></li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li><p>Check a CWL file for errors</p></li>
<li><p>Output debugging information</p></li>
<li><p>Interpret and fix commonly encountered error messages
keypoints:</p></li>
<li><p>Run the workflow with the <code>--validate</code> option to check
for errors</p></li>
<li><p>The <code>--debug</code> option will output more
information</p></li>
<li><p>‚ÄòWiring‚Äô errors won‚Äôt necessarily yield an error message</p></li>
</ul>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="a-firm-reality-check">A Firm Reality Check<a class="anchor" aria-label="anchor" href="#a-firm-reality-check"></a>
</h3>
<p>When working on a CWL workflow, you will probably encounter errors.
There are many different ways for errors to occur.</p>
<p>It is always very important to check the error message in the
terminal, because it will give you information on the error. This error
message will give you the type of error as well as the line of code that
contains the error.</p>
<p>We will showcase some of the common errors in this episode.</p>
<p>As a first step to check if your CWL script contains any errors, you
can run the workflow with the <code>--validate</code> flag.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">cwltool</span> <span class="at">--validate</span> /path/to/cwl_script.cwl</span></code></pre>
</div>
<p>It is possible for a valid script to still generate an error.</p>
<p>If you encounter an error, the best practice is to re-run the
workflow with the <code>--debug</code> flag. This will provide you with
extensive information on the error you encounter.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">cwltool</span> <span class="at">--debug</span> /path/to/cwl_script.cwl /path/to/cwl_input.yaml</span></code></pre>
</div>
</div>
<section id="syntax-errors"><h2 class="section-heading">Syntax Errors<a class="anchor" aria-label="anchor" href="#syntax-errors"></a>
</h2>
<hr class="half-width">
<p>When writing a piece of code, it is very easy to make a mistake in
your YAML syntax.</p>
<p>Some very common YAML errors are:</p>
<div class="section level3">
<h3 id="tabs">Tabs<a class="anchor" aria-label="anchor" href="#tabs"></a>
</h3>
<p>Using tabs instead of spaces. In YAML files indentations are made
using spaces, not tabs. Please download and run <a href="../files/debug/tab-error.cwl">this example</a> which includes a tab
character.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">$</span> cwltool tab-error.cwl workflow_input.yml</span></code></pre>
</div>
<div id="accordionSpoiler1" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler1" aria-expanded="false" aria-controls="collapseSpoiler1">
  <h3 class="accordion-header" id="headingSpoiler1">
<div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div>Show details</h3>
</button>
<div id="collapseSpoiler1" class="accordion-collapse collapse" data-bs-parent="#accordionSpoiler1" aria-labelledby="headingSpoiler1">
<div class="accordion-body">
<pre><code>ERROR Tool definition failed validation:
while scanning for the next token
file:///tab-error.cwl:5:1:   found character '\t' that cannot start any token</code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section id="field-name-typos"><h2 class="section-heading">Field Name Typos<a class="anchor" aria-label="anchor" href="#field-name-typos"></a>
</h2>
<hr class="half-width">
<p>Typos in field names. It is very easy to forget for example the
capital letters in field names.</p>
<p>Errors with typos in field names will show
<code>invalid field</code>.</p>
<p><strong>rna_seq_workflow_fieldname_fail.cwl</strong></p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">YAML<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode yaml" tabindex="0"><code class="sourceCode yaml"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">cwlVersion</span><span class="kw">:</span><span class="at"> v1.2</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="fu">class</span><span class="kw">:</span><span class="at"> Workflow</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="fu">inputs</span><span class="kw">:</span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="at">  </span><span class="fu">rna_reads_fruitfly</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="at">  </span><span class="fu">ref_fruitfly_genome</span><span class="kw">:</span><span class="at"> Directory</span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="fu">steps</span><span class="kw">:</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="at">  </span><span class="fu">quality_control</span><span class="kw">:</span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span><span class="at"> bio-cwl-tools/fastqc/fastqc_2.cwl</span></span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="at">      </span><span class="fu">reads_file</span><span class="kw">:</span><span class="at"> rna_reads_fruitfly</span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at">html_file</span><span class="kw">]</span></span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a></span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a><span class="at">  </span><span class="fu">mapping_reads</span><span class="kw">:</span></span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a><span class="at">    </span><span class="fu">requirements</span><span class="kw">:</span></span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a><span class="at">      </span><span class="fu">ResourceRequirement</span><span class="kw">:</span></span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a><span class="at">        </span><span class="fu">ramMin</span><span class="kw">:</span><span class="at"> </span><span class="dv">5120</span></span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span><span class="at"> bio-cwl-tools/STAR/STAR-Align.cwl</span></span>
<span id="cb5-20"><a href="#cb5-20" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb5-21"><a href="#cb5-21" tabindex="-1"></a><span class="at">      </span><span class="fu">RunThreadN</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="fu">default</span><span class="kw">:</span><span class="at"> </span><span class="dv">4</span><span class="kw">}</span></span>
<span id="cb5-22"><a href="#cb5-22" tabindex="-1"></a><span class="at">      </span><span class="fu">GenomeDir</span><span class="kw">:</span><span class="at"> ref_fruitfly_genome</span></span>
<span id="cb5-23"><a href="#cb5-23" tabindex="-1"></a><span class="at">      </span><span class="fu">ForwardReads</span><span class="kw">:</span><span class="at"> rna_reads_fruitfly</span></span>
<span id="cb5-24"><a href="#cb5-24" tabindex="-1"></a><span class="at">      </span><span class="fu">OutSAMtype</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="fu">default</span><span class="kw">:</span><span class="at"> BAM</span><span class="kw">}</span></span>
<span id="cb5-25"><a href="#cb5-25" tabindex="-1"></a><span class="at">      </span><span class="fu">SortedByCoordinate</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="fu">default</span><span class="kw">:</span><span class="at"> </span><span class="ch">true</span><span class="kw">}</span></span>
<span id="cb5-26"><a href="#cb5-26" tabindex="-1"></a><span class="at">      </span><span class="fu">OutSAMunmapped</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="fu">default</span><span class="kw">:</span><span class="at"> Within</span><span class="kw">}</span></span>
<span id="cb5-27"><a href="#cb5-27" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at">alignment</span><span class="kw">]</span></span>
<span id="cb5-28"><a href="#cb5-28" tabindex="-1"></a></span>
<span id="cb5-29"><a href="#cb5-29" tabindex="-1"></a><span class="at">  </span><span class="fu">index_alignment</span><span class="kw">:</span></span>
<span id="cb5-30"><a href="#cb5-30" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span><span class="at"> bio-cwl-tools/samtools/samtools_index.cwl</span></span>
<span id="cb5-31"><a href="#cb5-31" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb5-32"><a href="#cb5-32" tabindex="-1"></a><span class="at">      </span><span class="fu">bam_sorted</span><span class="kw">:</span><span class="at"> mapping_reads/alignment</span></span>
<span id="cb5-33"><a href="#cb5-33" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at">bam_sorted_indexed</span><span class="kw">]</span></span>
<span id="cb5-34"><a href="#cb5-34" tabindex="-1"></a></span>
<span id="cb5-35"><a href="#cb5-35" tabindex="-1"></a><span class="fu">outputs</span><span class="kw">:</span></span>
<span id="cb5-36"><a href="#cb5-36" tabindex="-1"></a><span class="at">  </span><span class="fu">qc_html</span><span class="kw">:</span></span>
<span id="cb5-37"><a href="#cb5-37" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb5-38"><a href="#cb5-38" tabindex="-1"></a><span class="at">    </span><span class="fu">outputsource</span><span class="kw">:</span><span class="at"> quality_control/html_file</span></span>
<span id="cb5-39"><a href="#cb5-39" tabindex="-1"></a><span class="at">  </span><span class="fu">bam_sorted_indexed</span><span class="kw">:</span></span>
<span id="cb5-40"><a href="#cb5-40" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb5-41"><a href="#cb5-41" tabindex="-1"></a><span class="at">    </span><span class="fu">outputSource</span><span class="kw">:</span><span class="at"> index_alignment/bam_sorted_indexed</span></span></code></pre>
</div>
<p><strong>Validate command</strong></p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="ex">cwltool</span> <span class="at">--validate</span> rna_seq_workflow_fieldname_fail.cwl</span></code></pre>
</div>
<div id="accordionSpoiler2" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler2" aria-expanded="false" aria-controls="collapseSpoiler2">
  <h3 class="accordion-header" id="headingSpoiler2">
<div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div>CWLTOOL VALIDATE ERROR MESSAGE</h3>
</button>
<div id="collapseSpoiler2" class="accordion-collapse collapse" data-bs-parent="#accordionSpoiler2" aria-labelledby="headingSpoiler2">
<div class="accordion-body">
<pre><code>ERROR Tool definition failed validation:
rna_seq_workflow_fieldname_fail.cwl:1:1:   Object `rna_seq_workflow_fieldname_fail.cwl` is not valid
                                          because
                                            tried `Workflow` but
rna_seq_workflow_fieldname_fail.cwl:35:1:     the `outputs` field is not valid because
rna_seq_workflow_fieldname_fail.cwl:36:3:       item is invalid because
rna_seq_workflow_fieldname_fail.cwl:38:5:         invalid field `outputsource`, expected one of:
                                                  'label', 'secondaryFiles', 'streamable', 'doc', 'id',
                                                  'format', 'outputSource', 'linkMerge', 'pickValue', 'type'</code></pre>
</div>
</div>
</div>
</div>
<div id="ides-are-your-friend" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="ides-are-your-friend" class="callout-inner">
<h3 class="callout-title">IDEs are your friend<a class="anchor" aria-label="anchor" href="#ides-are-your-friend"></a>
</h3>
<div class="callout-content">
<p>Using an IDE can help warn of incorrect fields before needing to
validate via the command-line tool</p>
</div>
</div>
</div>
</section><section id="variable-name-typos"><h2 class="section-heading">Variable Name Typos<a class="anchor" aria-label="anchor" href="#variable-name-typos"></a>
</h2>
<hr class="half-width">
<p>Typos in variable names.</p>
<p>Similar to typos in field names, it is easy to make a mistake in
referencing to a variable.<br>
These errors will show
<code>Field references unknown identifier.</code></p>
<p><strong>rna_seq_workflow_varname_fail.cwl</strong></p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">YAML<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode yaml" tabindex="0"><code class="sourceCode yaml"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="fu">cwlVersion</span><span class="kw">:</span><span class="at"> v1.2</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="fu">class</span><span class="kw">:</span><span class="at"> Workflow</span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="fu">inputs</span><span class="kw">:</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="at">  </span><span class="fu">rna_reads_fruitfly</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="at">  </span><span class="fu">ref_fruitfly_genome</span><span class="kw">:</span><span class="at"> Directory</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="fu">steps</span><span class="kw">:</span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="at">  </span><span class="fu">quality_control</span><span class="kw">:</span></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span><span class="at"> bio-cwl-tools/fastqc/fastqc_2.cwl</span></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a><span class="at">      </span><span class="fu">reads_file</span><span class="kw">:</span><span class="at"> rna_reads_fruitfly</span></span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at">html_file</span><span class="kw">]</span></span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a></span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a><span class="at">  </span><span class="fu">mapping_reads</span><span class="kw">:</span></span>
<span id="cb8-16"><a href="#cb8-16" tabindex="-1"></a><span class="at">    </span><span class="fu">requirements</span><span class="kw">:</span></span>
<span id="cb8-17"><a href="#cb8-17" tabindex="-1"></a><span class="at">      </span><span class="fu">ResourceRequirement</span><span class="kw">:</span></span>
<span id="cb8-18"><a href="#cb8-18" tabindex="-1"></a><span class="at">        </span><span class="fu">ramMin</span><span class="kw">:</span><span class="at"> </span><span class="dv">5120</span></span>
<span id="cb8-19"><a href="#cb8-19" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span><span class="at"> bio-cwl-tools/STAR/STAR-Align.cwl</span></span>
<span id="cb8-20"><a href="#cb8-20" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb8-21"><a href="#cb8-21" tabindex="-1"></a><span class="at">      </span><span class="fu">RunThreadN</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="fu">default</span><span class="kw">:</span><span class="at"> </span><span class="dv">4</span><span class="kw">}</span></span>
<span id="cb8-22"><a href="#cb8-22" tabindex="-1"></a><span class="at">      </span><span class="fu">GenomeDir</span><span class="kw">:</span><span class="at"> ref_fruitfly_genome</span></span>
<span id="cb8-23"><a href="#cb8-23" tabindex="-1"></a><span class="at">      </span><span class="fu">ForwardReads</span><span class="kw">:</span><span class="at"> rna_reads_fruitfly</span></span>
<span id="cb8-24"><a href="#cb8-24" tabindex="-1"></a><span class="at">      </span><span class="fu">OutSAMtype</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="fu">default</span><span class="kw">:</span><span class="at"> BAM</span><span class="kw">}</span></span>
<span id="cb8-25"><a href="#cb8-25" tabindex="-1"></a><span class="at">      </span><span class="fu">SortedByCoordinate</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="fu">default</span><span class="kw">:</span><span class="at"> </span><span class="ch">true</span><span class="kw">}</span></span>
<span id="cb8-26"><a href="#cb8-26" tabindex="-1"></a><span class="at">      </span><span class="fu">OutSAMunmapped</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="fu">default</span><span class="kw">:</span><span class="at"> Within</span><span class="kw">}</span></span>
<span id="cb8-27"><a href="#cb8-27" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at">alignment</span><span class="kw">]</span></span>
<span id="cb8-28"><a href="#cb8-28" tabindex="-1"></a></span>
<span id="cb8-29"><a href="#cb8-29" tabindex="-1"></a><span class="at">  </span><span class="fu">index_alignment</span><span class="kw">:</span></span>
<span id="cb8-30"><a href="#cb8-30" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span><span class="at"> bio-cwl-tools/samtools/samtools_index.cwl</span></span>
<span id="cb8-31"><a href="#cb8-31" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb8-32"><a href="#cb8-32" tabindex="-1"></a><span class="at">      </span><span class="fu">bam_sorted</span><span class="kw">:</span><span class="at"> mapping_reads/alignments</span></span>
<span id="cb8-33"><a href="#cb8-33" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at">bam_sorted_indexed</span><span class="kw">]</span></span>
<span id="cb8-34"><a href="#cb8-34" tabindex="-1"></a></span>
<span id="cb8-35"><a href="#cb8-35" tabindex="-1"></a><span class="fu">outputs</span><span class="kw">:</span></span>
<span id="cb8-36"><a href="#cb8-36" tabindex="-1"></a><span class="at">  </span><span class="fu">qc_html</span><span class="kw">:</span></span>
<span id="cb8-37"><a href="#cb8-37" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb8-38"><a href="#cb8-38" tabindex="-1"></a><span class="at">    </span><span class="fu">outputSource</span><span class="kw">:</span><span class="at"> quality_control/html_file</span></span>
<span id="cb8-39"><a href="#cb8-39" tabindex="-1"></a><span class="at">  </span><span class="fu">bam_sorted_indexed</span><span class="kw">:</span></span>
<span id="cb8-40"><a href="#cb8-40" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb8-41"><a href="#cb8-41" tabindex="-1"></a><span class="at">    </span><span class="fu">outputSource</span><span class="kw">:</span><span class="at"> index_alignment/bam_sorted_indexed</span></span></code></pre>
</div>
<p><strong>Validate command</strong></p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="ex">$</span> cwltool <span class="at">--validate</span> rna_seq_workflow_varname_fail.cwl</span></code></pre>
</div>
<div id="accordionSpoiler3" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler3" aria-expanded="false" aria-controls="collapseSpoiler3">
  <h3 class="accordion-header" id="headingSpoiler3">
<div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div>CWLTOOL VALIDATE ERROR MESSAGE</h3>
</button>
<div id="collapseSpoiler3" class="accordion-collapse collapse" data-bs-parent="#accordionSpoiler3" aria-labelledby="headingSpoiler3">
<div class="accordion-body">
<pre><code>ERROR Tool definition failed validation:
rna_seq_workflow_varname_fail.cwl:8:1:  checking field `steps`
rna_seq_workflow_varname_fail.cwl:29:3:   checking object
                                          `rna_seq_workflow_varname_fail.cwl#index_alignment`
rna_seq_workflow_varname_fail.cwl:31:5:     checking field `in`
rna_seq_workflow_varname_fail.cwl:32:7:       checking object
                                              `rna_seq_workflow_varname_fail.cwl#index_alignment/bam_sorted`
                                                Field `source` references unknown identifier
                                                `mapping_reads/alignments`, tried
                                                file:///.../rna_seq_workflow_varname_fail.cwl#mapping_reads/alignments</code></pre>
</div>
</div>
</div>
</div>
<div id="ides-are-your-friend-1" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="ides-are-your-friend-1" class="callout-inner">
<h3 class="callout-title">IDEs are your friend<a class="anchor" aria-label="anchor" href="#ides-are-your-friend-1"></a>
</h3>
<div class="callout-content">
<p>Using an IDE, a simple Ctrl+F on a variable can help you see where
that variable is present throughout a CWL code. Only one occurence of a
variable might mean it has been spelt differently elsewhere.</p>
</div>
</div>
</div>
</section><section id="wiring-error"><h2 class="section-heading">Wiring error<a class="anchor" aria-label="anchor" href="#wiring-error"></a>
</h2>
<hr class="half-width">
<p>Wiring errors often occur when you forget to add an output from a
workflow‚Äôs step to the <code>outputs</code> section.</p>
<p>This doesn‚Äôt cause an error message, but there won‚Äôt be any output in
your directory. To get the desired output you have to run the workflow
again.</p>
<p>Best practice is to check your <code>outputs</code> section before
running your script to make sure all the outputs you want are there.</p>
<div id="output-collections" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="output-collections" class="callout-inner">
<h3 class="callout-title">Output Collections<a class="anchor" aria-label="anchor" href="#output-collections"></a>
</h3>
<div class="callout-content">
<p>All file / directory outputs of a workflow or tool will be placed
into a single directory.</p>
<p>Ensure all expected files and directories are there. For string /
boolean output, splitting stderr and stdout of the cwltool commandline
into separate files allows the user to easily look through the stdout of
a workflow without needing the noise of stderr. Redirects are briefly
discussed in the second episode of this tutorial.</p>
</div>
</div>
</div>
</section><section id="type-mismatch"><h2 class="section-heading">Type mismatch<a class="anchor" aria-label="anchor" href="#type-mismatch"></a>
</h2>
<hr class="half-width">
<p>Type errors take place when there is a mismatch in type between
variables. When you declare a variable in the <code>inputs</code>
section, the type of this variable has to match the type in the YAML
inputs file and the type used in one of the workflows steps. The error
message that is shown when this error occurs will tell you on which line
the mismatch happens.</p>
<p><strong>rna_seq_workflow_type_fail.cwl</strong></p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">YAML<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode yaml" tabindex="0"><code class="sourceCode yaml"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="fu">cwlVersion</span><span class="kw">:</span><span class="at"> v1.2</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="fu">class</span><span class="kw">:</span><span class="at"> Workflow</span></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a><span class="fu">inputs</span><span class="kw">:</span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a><span class="at">  </span><span class="fu">rna_reads_fruitfly</span><span class="kw">:</span><span class="at"> int</span></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a><span class="at">  </span><span class="fu">ref_fruitfly_genome</span><span class="kw">:</span><span class="at"> Directory</span></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a></span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a><span class="fu">steps</span><span class="kw">:</span></span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a><span class="at">  </span><span class="fu">quality_control</span><span class="kw">:</span></span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span><span class="at"> bio-cwl-tools/fastqc/fastqc_2.cwl</span></span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a><span class="at">      </span><span class="fu">reads_file</span><span class="kw">:</span><span class="at"> rna_reads_fruitfly</span></span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at">html_file</span><span class="kw">]</span></span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a></span>
<span id="cb11-15"><a href="#cb11-15" tabindex="-1"></a><span class="at">  </span><span class="fu">mapping_reads</span><span class="kw">:</span></span>
<span id="cb11-16"><a href="#cb11-16" tabindex="-1"></a><span class="at">    </span><span class="fu">requirements</span><span class="kw">:</span></span>
<span id="cb11-17"><a href="#cb11-17" tabindex="-1"></a><span class="at">      </span><span class="fu">ResourceRequirement</span><span class="kw">:</span></span>
<span id="cb11-18"><a href="#cb11-18" tabindex="-1"></a><span class="at">        </span><span class="fu">ramMin</span><span class="kw">:</span><span class="at"> </span><span class="dv">5120</span></span>
<span id="cb11-19"><a href="#cb11-19" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span><span class="at"> bio-cwl-tools/STAR/STAR-Align.cwl</span></span>
<span id="cb11-20"><a href="#cb11-20" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb11-21"><a href="#cb11-21" tabindex="-1"></a><span class="at">      </span><span class="fu">RunThreadN</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="fu">default</span><span class="kw">:</span><span class="at"> </span><span class="dv">4</span><span class="kw">}</span></span>
<span id="cb11-22"><a href="#cb11-22" tabindex="-1"></a><span class="at">      </span><span class="fu">GenomeDir</span><span class="kw">:</span><span class="at"> ref_fruitfly_genome</span></span>
<span id="cb11-23"><a href="#cb11-23" tabindex="-1"></a><span class="at">      </span><span class="fu">ForwardReads</span><span class="kw">:</span><span class="at"> rna_reads_fruitfly</span></span>
<span id="cb11-24"><a href="#cb11-24" tabindex="-1"></a><span class="at">      </span><span class="fu">OutSAMtype</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="fu">default</span><span class="kw">:</span><span class="at"> BAM</span><span class="kw">}</span></span>
<span id="cb11-25"><a href="#cb11-25" tabindex="-1"></a><span class="at">      </span><span class="fu">SortedByCoordinate</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="fu">default</span><span class="kw">:</span><span class="at"> </span><span class="ch">true</span><span class="kw">}</span></span>
<span id="cb11-26"><a href="#cb11-26" tabindex="-1"></a><span class="at">      </span><span class="fu">OutSAMunmapped</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="fu">default</span><span class="kw">:</span><span class="at"> Within</span><span class="kw">}</span></span>
<span id="cb11-27"><a href="#cb11-27" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at">alignment</span><span class="kw">]</span></span>
<span id="cb11-28"><a href="#cb11-28" tabindex="-1"></a></span>
<span id="cb11-29"><a href="#cb11-29" tabindex="-1"></a><span class="at">  </span><span class="fu">index_alignment</span><span class="kw">:</span></span>
<span id="cb11-30"><a href="#cb11-30" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span><span class="at"> bio-cwl-tools/samtools/samtools_index.cwl</span></span>
<span id="cb11-31"><a href="#cb11-31" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb11-32"><a href="#cb11-32" tabindex="-1"></a><span class="at">      </span><span class="fu">bam_sorted</span><span class="kw">:</span><span class="at"> mapping_reads/alignment</span></span>
<span id="cb11-33"><a href="#cb11-33" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at">bam_sorted_indexed</span><span class="kw">]</span></span>
<span id="cb11-34"><a href="#cb11-34" tabindex="-1"></a></span>
<span id="cb11-35"><a href="#cb11-35" tabindex="-1"></a><span class="fu">outputs</span><span class="kw">:</span></span>
<span id="cb11-36"><a href="#cb11-36" tabindex="-1"></a><span class="at">  </span><span class="fu">qc_html</span><span class="kw">:</span></span>
<span id="cb11-37"><a href="#cb11-37" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb11-38"><a href="#cb11-38" tabindex="-1"></a><span class="at">    </span><span class="fu">outputSource</span><span class="kw">:</span><span class="at"> quality_control/html_file</span></span>
<span id="cb11-39"><a href="#cb11-39" tabindex="-1"></a><span class="at">  </span><span class="fu">bam_sorted_indexed</span><span class="kw">:</span></span>
<span id="cb11-40"><a href="#cb11-40" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb11-41"><a href="#cb11-41" tabindex="-1"></a><span class="at">    </span><span class="fu">outputSource</span><span class="kw">:</span><span class="at"> index_alignment/bam_sorted_indexed</span></span></code></pre>
</div>
<p><strong>Validation Command</strong></p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="ex">$</span> cwltool rna_seq_workflow_type_fail.cwl workflow_input_debug.yml</span></code></pre>
</div>
<div id="accordionSpoiler4" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler4" aria-expanded="false" aria-controls="collapseSpoiler4">
  <h3 class="accordion-header" id="headingSpoiler4">
<div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div>Incompatible type cwltool error message</h3>
</button>
<div id="collapseSpoiler4" class="accordion-collapse collapse" data-bs-parent="#accordionSpoiler4" aria-labelledby="headingSpoiler4">
<div class="accordion-body">
<pre><code>rna_seq_workflow_type_fail.cwl:5:3:   Source 'rna_reads_fruitfly' of type "int" is incompatible
rna_seq_workflow_type_fail.cwl:12:7:   with sink 'reads_file' of type "File"
rna_seq_workflow_type_fail.cwl:5:3:   Source 'rna_reads_fruitfly' of type "int" is incompatible
rna_seq_workflow_type_fail.cwl:23:7:   with sink 'ForwardReads' of type ["File", {"type":
                                       "array", "items": "File"}]
</code></pre>
</div>
</div>
</div>
</div>
</section><section id="format-error"><h2 class="section-heading">Format error<a class="anchor" aria-label="anchor" href="#format-error"></a>
</h2>
<hr class="half-width">
<p>Some files need a specific format that needs to be specified in the
YAML inputs file, for example the fastq file in the RNA-seq
analysis.</p>
<p>When you don‚Äôt specify a format, an error will occur. You can for
example use the <a href="https://www.ebi.ac.uk/ols/ontologies/edam" class="external-link">EDAM</a> ontology.</p>
<div id="format-requirements" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="format-requirements" class="callout-inner">
<h3 class="callout-title">Format requirements<a class="anchor" aria-label="anchor" href="#format-requirements"></a>
</h3>
<div class="callout-content">
<p>The format attribute for a File entry is only required if the format
attribute is specified on the workflow input.</p>
<p>You may use
<code>cwltool --make-template /path/to/cwl_workflow.cwl</code> to set
the formats for each input for you.</p>
</div>
</div>
</div>
<p><strong>rna_seq_workflow_with_format.cwl</strong></p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">YAML<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode yaml" tabindex="0"><code class="sourceCode yaml"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="fu">cwlVersion</span><span class="kw">:</span><span class="at"> v1.2</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="fu">class</span><span class="kw">:</span><span class="at"> Workflow</span></span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a><span class="fu">inputs</span><span class="kw">:</span></span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a><span class="at">  </span><span class="fu">rna_reads_fruitfly</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a><span class="at">  </span><span class="fu">ref_fruitfly_genome</span><span class="kw">:</span><span class="at"> Directory</span></span>
<span id="cb14-7"><a href="#cb14-7" tabindex="-1"></a></span>
<span id="cb14-8"><a href="#cb14-8" tabindex="-1"></a><span class="fu">steps</span><span class="kw">:</span></span>
<span id="cb14-9"><a href="#cb14-9" tabindex="-1"></a><span class="at">  </span><span class="fu">quality_control</span><span class="kw">:</span></span>
<span id="cb14-10"><a href="#cb14-10" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span><span class="at"> bio-cwl-tools/fastqc/fastqc_2.cwl</span></span>
<span id="cb14-11"><a href="#cb14-11" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb14-12"><a href="#cb14-12" tabindex="-1"></a><span class="at">      </span><span class="fu">reads_file</span><span class="kw">:</span><span class="at"> rna_reads_fruitfly</span></span>
<span id="cb14-13"><a href="#cb14-13" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at">html_file</span><span class="kw">]</span></span>
<span id="cb14-14"><a href="#cb14-14" tabindex="-1"></a></span>
<span id="cb14-15"><a href="#cb14-15" tabindex="-1"></a><span class="at">  </span><span class="fu">mapping_reads</span><span class="kw">:</span></span>
<span id="cb14-16"><a href="#cb14-16" tabindex="-1"></a><span class="at">    </span><span class="fu">requirements</span><span class="kw">:</span></span>
<span id="cb14-17"><a href="#cb14-17" tabindex="-1"></a><span class="at">      </span><span class="fu">ResourceRequirement</span><span class="kw">:</span></span>
<span id="cb14-18"><a href="#cb14-18" tabindex="-1"></a><span class="at">        </span><span class="fu">ramMin</span><span class="kw">:</span><span class="at"> </span><span class="dv">5120</span></span>
<span id="cb14-19"><a href="#cb14-19" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span><span class="at"> bio-cwl-tools/STAR/STAR-Align.cwl</span></span>
<span id="cb14-20"><a href="#cb14-20" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb14-21"><a href="#cb14-21" tabindex="-1"></a><span class="at">      </span><span class="fu">RunThreadN</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="fu">default</span><span class="kw">:</span><span class="at"> </span><span class="dv">4</span><span class="kw">}</span></span>
<span id="cb14-22"><a href="#cb14-22" tabindex="-1"></a><span class="at">      </span><span class="fu">GenomeDir</span><span class="kw">:</span><span class="at"> ref_fruitfly_genome</span></span>
<span id="cb14-23"><a href="#cb14-23" tabindex="-1"></a><span class="at">      </span><span class="fu">ForwardReads</span><span class="kw">:</span><span class="at"> rna_reads_fruitfly</span></span>
<span id="cb14-24"><a href="#cb14-24" tabindex="-1"></a><span class="at">      </span><span class="fu">OutSAMtype</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="fu">default</span><span class="kw">:</span><span class="at"> BAM</span><span class="kw">}</span></span>
<span id="cb14-25"><a href="#cb14-25" tabindex="-1"></a><span class="at">      </span><span class="fu">SortedByCoordinate</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="fu">default</span><span class="kw">:</span><span class="at"> </span><span class="ch">true</span><span class="kw">}</span></span>
<span id="cb14-26"><a href="#cb14-26" tabindex="-1"></a><span class="at">      </span><span class="fu">OutSAMunmapped</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span><span class="fu">default</span><span class="kw">:</span><span class="at"> Within</span><span class="kw">}</span></span>
<span id="cb14-27"><a href="#cb14-27" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at">alignment</span><span class="kw">]</span></span>
<span id="cb14-28"><a href="#cb14-28" tabindex="-1"></a></span>
<span id="cb14-29"><a href="#cb14-29" tabindex="-1"></a><span class="at">  </span><span class="fu">index_alignment</span><span class="kw">:</span></span>
<span id="cb14-30"><a href="#cb14-30" tabindex="-1"></a><span class="at">    </span><span class="fu">run</span><span class="kw">:</span><span class="at"> bio-cwl-tools/samtools/samtools_index.cwl</span></span>
<span id="cb14-31"><a href="#cb14-31" tabindex="-1"></a><span class="at">    </span><span class="fu">in</span><span class="kw">:</span></span>
<span id="cb14-32"><a href="#cb14-32" tabindex="-1"></a><span class="at">      </span><span class="fu">bam_sorted</span><span class="kw">:</span><span class="at"> mapping_reads/alignment</span></span>
<span id="cb14-33"><a href="#cb14-33" tabindex="-1"></a><span class="at">    </span><span class="fu">out</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at">bam_sorted_indexed</span><span class="kw">]</span></span>
<span id="cb14-34"><a href="#cb14-34" tabindex="-1"></a></span>
<span id="cb14-35"><a href="#cb14-35" tabindex="-1"></a><span class="fu">outputs</span><span class="kw">:</span></span>
<span id="cb14-36"><a href="#cb14-36" tabindex="-1"></a><span class="at">  </span><span class="fu">qc_html</span><span class="kw">:</span></span>
<span id="cb14-37"><a href="#cb14-37" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb14-38"><a href="#cb14-38" tabindex="-1"></a><span class="at">    </span><span class="fu">outputSource</span><span class="kw">:</span><span class="at"> quality_control/html_file</span></span>
<span id="cb14-39"><a href="#cb14-39" tabindex="-1"></a><span class="at">  </span><span class="fu">bam_sorted_indexed</span><span class="kw">:</span></span>
<span id="cb14-40"><a href="#cb14-40" tabindex="-1"></a><span class="at">    </span><span class="fu">type</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb14-41"><a href="#cb14-41" tabindex="-1"></a><span class="at">    </span><span class="fu">outputSource</span><span class="kw">:</span><span class="at"> index_alignment/bam_sorted_indexed</span></span></code></pre>
</div>
<p><strong>workflow_input_undefined_format.yaml</strong></p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">YAML<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode yaml" tabindex="0"><code class="sourceCode yaml"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="fu">rna_reads_fruitfly</span><span class="kw">:</span></span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="at">  </span><span class="fu">class</span><span class="kw">:</span><span class="at"> File</span></span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a><span class="at">  </span><span class="fu">location</span><span class="kw">:</span><span class="at"> rnaseq/GSM461177_1_subsampled.fastqsanger</span></span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a><span class="fu">ref_fruitfly_genome</span><span class="kw">:</span></span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a><span class="at">  </span><span class="fu">class</span><span class="kw">:</span><span class="at"> Directory</span></span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a><span class="at">  </span><span class="fu">location</span><span class="kw">:</span><span class="at"> rnaseq/dm6-STAR-index</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="ex">$</span> cwltool rna_seq_workflow_with_format.cwl workflow_input_undefined.yml</span></code></pre>
</div>
<div id="accordionSpoiler5" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler5" aria-expanded="false" aria-controls="collapseSpoiler5">
  <h3 class="accordion-header" id="headingSpoiler5">
<div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div>Incompatible format error message</h3>
</button>
<div id="collapseSpoiler5" class="accordion-collapse collapse" data-bs-parent="#accordionSpoiler5" aria-labelledby="headingSpoiler5">
<div class="accordion-body">
<pre><code>ERROR Exception on step 'mapping_reads'
ERROR [step mapping_reads] Cannot make job: Expected value of 'ForwardReads' to have format http://edamontology.org/format_1930 but
  File has no 'format' defined: {
    "class": "File",
    "location": "file:///.../rnaseq/GSM461177_1_subsampled.fastqsanger",
    "size": 142867948,
    "basename": "GSM461177_1_subsampled.fastqsanger",
    "nameroot": "GSM461177_1_subsampled",
    "nameext": ".fastqsanger"
}</code></pre>
</div>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></section><section id="aio-more_info"><p>Content from <a href="more_info.html">More Information</a></p>
<hr>
<p>Last updated on 2024-05-14 |
        
        <a href="https://github.com/__USERNAME__/__PROJECT__/edit/main/episodes/more_info.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 0 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<p>If you want to know more about CWL script and workflows, you can look
at one of these websites:</p>
<ul>
<li><a href="https://www.commonwl.org/user_guide/index.html" class="external-link">CWL User
Guide</a></li>
<li><a href="https://www.commonwl.org/user_guide/topics/yaml-guide" class="external-link">YAML
Guide</a></li>
<li>Specs
<ul>
<li><a href="https://www.commonwl.org/v1.2/CommandLineTool.html" class="external-link">CWL
Command Line Tool</a></li>
<li><a href="https://www.commonwl.org/v1.2/Workflow.html" class="external-link">CWL
Workflow</a></li>
</ul>
</li>
<li>
<a href="https://www.commonwl.org/user_guide/topics/best-practices.html" class="external-link">Recommended
Practices</a> in CWL</li>
</ul></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/__USERNAME__/__PROJECT__/edit/main/README.md" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/__USERNAME__/__PROJECT__/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/__USERNAME__/__PROJECT__/" class="external-link">Source</a></p>
				<p><a href="https://github.com/__USERNAME__/__PROJECT__/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:https://www.commonwl.org/#Support" class="external-link">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.4" class="external-link">sandpaper (0.16.4)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.5" class="external-link">pegboard (0.7.5)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.2" class="external-link">varnish (1.0.2)</a></p>
			</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://__USERNAME__.github.io/__PROJECT__/instructor/aio.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, workflows, Common Workflow Language",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://__USERNAME__.github.io/__PROJECT__/instructor/aio.html",
  "identifier": "https://__USERNAME__.github.io/__PROJECT__/instructor/aio.html",
  "dateCreated": "2024-04-29",
  "dateModified": "2024-05-14",
  "datePublished": "2024-05-14"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code -->
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

