## Data Requirements

You will need to download some example files for this lesson. In this tutorial we will use RNA sequencing data.

### Setting up a practice Git repository

For this tutorial some existing tools are needed to build the workflow. These existing tools will be imported via GitHub.


#### Make a repository on GitHub.com

Head to [github.com/new][github_new_repo] and create a new repository.  

Name it `novice-tutorial-exercises`.  

Make sure the repository is `public`.  

Check 'Add a README file', and choose `GNU General Public License v3.0` as your license.

#### Clone the repository

```bash
git clone git@github.com/YOUR_GITHUB_USERNAME/novice-tutorial-exercises
```

Next, we need to move into our git repo:

```bash
cd novice-tutorial-exercises
```

<br>

#### Add in bio-cwl-tools as a submodule

Then import bio-cwl-tools with this command:

```bash
git submodule add https://github.com/common-workflow-library/bio-cwl-tools.git
```

<br>


### Downloading sample and reference data

Create a new directory inside the `novice-tutorial-exercises` directory and download the data:

:::::::::::::::: callout

### Using subshells

By running the following chunk in brackets the console will return to the
original working directory after the download is complete

::::::::::::::::

```bash
mkdir rnaseq
(
  cd rnaseq
  wget --quiet https://zenodo.org/record/4541751/files/GSM461177_1_subsampled.fastqsanger
  wget --quiet https://zenodo.org/record/4541751/files/GSM461177_2_subsampled.fastqsanger
  wget --quiet https://zenodo.org/record/4541751/files/Drosophila_melanogaster.BDGP6.87.gtf
  wget --quiet https://hgdownload.soe.ucsc.edu/goldenPath/dm6/bigZips/dm6.fa.gz
  gunzip dm6.fa.gz  # STAR index requires an uncompressed reference genome
)
```

<br>

### Ignore data files in Git

Git is not designed to track large data files.  
We should exclude the rnaseq folder from being tracked in this repository.  

```
echo 'rnaseq/' >> .gitignore
```

### Add .gitignore to git

`.gitignore` is a file that is tracked like all other GitHub commands.  

```
git add .gitignore
git commit -m "Don't track files in the rnaseq directory"
git push
```

<br>

### STAR Genome index

To run the STAR aligner tool, index files generated from the reference genome are needed.
The index files can be downloaded, or generated yourself from the unindexed reference genome.

These two options are detailed below -- choose the one most appropriate to your set-up.

:::::::::::::::::::: spoiler

#### Option 1: Download the index

At least 9 GB of memory is required to generate the index, which will occupy 3.3GB of disk.

If your computer doesn't have that much memory, then you can download the directory
by running the following in the `rnaseq` directory:

```bash
wget --quiet --output-document - https://dataverse.nl/api/access/datafile/266295 | \
tar -C rnaseq -x --xz
```

:::::::::::::::::::::

::::::::::::::::::::: spoiler

#### Option 2: Generate the index yourself

To generate the genome index yourself: create a new file named `dm6-star-index.yaml`
in the the `novice-tutorial-exercises` directory with the following contents:

```yaml
InputFiles:
 - class: File
   location: rnaseq/dm6.fa
   format: http://edamontology.org/format_1929  # FASTA
IndexName: 'dm6-STAR-index'
Overhang: 36
Gtf:
 class: File
 location: rnaseq/Drosophila_melanogaster.BDGP6.87.gtf
```

Next use the CWL reference runner `cwltool` that you installed above and the CWL description
for the indexing mode of the STAR aligner that was downloaded in the `bio-cwl-tools` directory
to index the genome and place the result in the `rnaseq` directory alongside the other files:

```
cwltool --outdir rnaseq/ bio-cwl-tools/STAR/STAR-Index.cwl dm6-star-index.yaml
```

It should take 10-15 minutes for the index to be generated.

::::::::::::::::::::: callout

### STAR Index Memory Requirements
To generate the genome index you will need at least 9 GB of RAM.

If you do not allocate enough RAM the tool will not crash, but the process will stick
on the following step:
```
... sorting Suffix Array chunks and saving them to disk...
```
If this step does not finish within 10 minutes then it is likely the process has failed,
and should be cancelled.

MacOS users can configure  Docker Desktop to allocate more memory
from the menu "Preferences" and then selecting "Resources".

Windows users can configure WSL 2 to allocate more memory by opening the PowerShell
and entering the following:

```bash
# turn off all wsl instances such as docker-desktop
wsl --shutdown

notepad "$env:USERPROFILE/.wslconfig"
```

In `.wslconfig` add the following

```text
[wsl2]
memory=9GB
```

Save the file and right-click the Docker icon in the notifications area (or System tray)
and then click "Restart Dockerâ€¦"

:::::::::::::::::::::

:::::::::::::::::::::

<br>

### Validating your dataset

Let's make sure the data we've downloaded is correct and in the right structure.  
We can do this with a [checksum][checksum_wiki].


#### Setting the checksum.  

Under the rnaseq folder create the file `rnaseq.md5` with the following contents.  

::::::::::::::::: spoiler

### rnaseq.md5

```bash
3c8458cf67d71c22c4c420dd37e23ef2  Drosophila_melanogaster.BDGP6.87.gtf
036b68a2a8fe8725d48fb5fd89e2b8b2  GSM461177_1_subsampled.fastqsanger
87b09607057743ecf3e38448630421c9  GSM461177_2_subsampled.fastqsanger
80f57c0a3537d2e4cd9f1748e7b7da91  dm6-STAR-index/Genome
e1a0b37b3cae8af4871cfb8c293a98cb  dm6-STAR-index/SA
eec51bc2096fbfd71f2cdb970529a6f9  dm6-STAR-index/SAindex
d8f7048f5f882af92c7fda0c711c70ad  dm6-STAR-index/chrLength.txt
d727129afa13f91ae22bd35e29afd0c0  dm6-STAR-index/chrName.txt
63645d2aab7fd4c9ed356e071d7b9c2a  dm6-STAR-index/chrNameLength.txt
f897e16cb9a8d9a4f1cf242938f97ac6  dm6-STAR-index/chrStart.txt
d5212320af99792dc517f1d4fa830ec1  dm6-STAR-index/exonGeTrInfo.tab
5ef2835f5c116f55dd20522567152bf8  dm6-STAR-index/exonInfo.tab
11aa105726dbb95e8f27714fdbd8e897  dm6-STAR-index/geneInfo.tab
faa97745744f087d06879800a531edea  dm6-STAR-index/sjdbInfo.txt
8cd08db8cb269461db29d9eb23ad690b  dm6-STAR-index/sjdbList.fromGTF.out.tab
fd0b412f47ca662ce68e08ca6939a2c7  dm6-STAR-index/sjdbList.out.tab
bdf26dee24f8cd1ca7b1bd854c882411  dm6-STAR-index/transcriptInfo.tab
5aadf7ccab5a6b674e76516bf75eaa09  dm6.fa
```

:::::::::::::::::::::::::::

The following code chunk may be of assistance. 
You may also use a text editor.  

:::::::::::::::::::::: spoiler

### make_md5.sh

```bash
(
  cd rnaseq && \
  {
     echo "3c8458cf67d71c22c4c420dd37e23ef2  Drosophila_melanogaster.BDGP6.87.gtf"
     echo "036b68a2a8fe8725d48fb5fd89e2b8b2  GSM461177_1_subsampled.fastqsanger"
     echo "87b09607057743ecf3e38448630421c9  GSM461177_2_subsampled.fastqsanger"
     echo "80f57c0a3537d2e4cd9f1748e7b7da91  dm6-STAR-index/Genome"
     echo "e1a0b37b3cae8af4871cfb8c293a98cb  dm6-STAR-index/SA"
     echo "eec51bc2096fbfd71f2cdb970529a6f9  dm6-STAR-index/SAindex"
     echo "d8f7048f5f882af92c7fda0c711c70ad  dm6-STAR-index/chrLength.txt"
     echo "d727129afa13f91ae22bd35e29afd0c0  dm6-STAR-index/chrName.txt"
     echo "63645d2aab7fd4c9ed356e071d7b9c2a  dm6-STAR-index/chrNameLength.txt"
     echo "f897e16cb9a8d9a4f1cf242938f97ac6  dm6-STAR-index/chrStart.txt"
     echo "d5212320af99792dc517f1d4fa830ec1  dm6-STAR-index/exonGeTrInfo.tab"
     echo "5ef2835f5c116f55dd20522567152bf8  dm6-STAR-index/exonInfo.tab"
     echo "11aa105726dbb95e8f27714fdbd8e897  dm6-STAR-index/geneInfo.tab"
     echo "faa97745744f087d06879800a531edea  dm6-STAR-index/sjdbInfo.txt"
     echo "8cd08db8cb269461db29d9eb23ad690b  dm6-STAR-index/sjdbList.fromGTF.out.tab"
     echo "fd0b412f47ca662ce68e08ca6939a2c7  dm6-STAR-index/sjdbList.out.tab"
     echo "bdf26dee24f8cd1ca7b1bd854c882411  dm6-STAR-index/transcriptInfo.tab"
     echo "5aadf7ccab5a6b674e76516bf75eaa09  dm6.fa"
  } >> rnaseq.md5
)
```

::::::::::::::::::::::::



#### Checking your files


```
(
  cd rnaseq && \
  md5sum -c rnaseq.md5
)
```

::::::::::::::::::::: spoiler

### md5sum check command outputs

We expect the following outputs from this command:

```
Drosophila_melanogaster.BDGP6.87.gtf: OK
GSM461177_1_subsampled.fastqsanger: OK
GSM461177_2_subsampled.fastqsanger: OK
dm6-STAR-index/Genome: OK
dm6-STAR-index/Log.out: OK
dm6-STAR-index/SA: OK
dm6-STAR-index/SAindex: OK
dm6-STAR-index/chrLength.txt: OK
dm6-STAR-index/chrName.txt: OK
dm6-STAR-index/chrNameLength.txt: OK
dm6-STAR-index/chrStart.txt: OK
dm6-STAR-index/exonGeTrInfo.tab: OK
dm6-STAR-index/exonInfo.tab: OK
dm6-STAR-index/geneInfo.tab: OK
dm6-STAR-index/genomeParameters.txt: OK
dm6-STAR-index/sjdbInfo.txt: OK
dm6-STAR-index/sjdbList.fromGTF.out.tab: OK
dm6-STAR-index/sjdbList.out.tab: OK
dm6-STAR-index/transcriptInfo.tab: OK
dm6.fa: OK
```

::::::::::::::::::


::::::::::::::::::::: callout

### md5 Troubleshooting: Windows users only

If you see an error such as 

```
md5sum: 'Drosophila_melanogaster.BDGP6.87.gtf'$'\r': No such file or directory
: FAILED open or reader.BDGP6.87.gtf
```

it may be because `rnaseq.md5` has windows based line endings.  

The following code will install the [dos2unix][dos2unix] cli tool and then convert
the `rnaseq.md5` file to unix-based line endings.  

```bash
sudo apt update -yq
sudo apt install dos2unix -yq

dos2unix rnaseq/rnaseq.md5
```

Then retry the `md5sum` command

::::::::::::::::::::::::::


[dos2unix]: https://phoenixnap.com/kb/convert-dos-to-unix
[checksum_wiki]: https://en.wikipedia.org/wiki/Checksum
[github_new_repo]: https://github.com/new
