## Data Requirements

You will need to download some example files for this lesson. In this tutorial we will use RNA sequencing data.

### Setting up a practice Git repository

For this tutorial some existing tools are needed to build the workflow. These existing tools will be imported via GitHub.
First we need to create an empty git repository for all our files. To do this, use this command:

```bash
git init novice-tutorial-exercises
```

Next, we need to move into our empty git repo:

```bash
cd novice-tutorial-exercises
```

Then import bio-cwl-tools with this command:

```bash
git submodule add https://github.com/common-workflow-library/bio-cwl-tools.git
```

### Downloading sample and reference data
Create a new directory inside the `novice-tutorial-exercises` directory and download the data:

:::::::::::::::: callout

By running the following chunk in brackets the console will return to the
original working directory after the download is complete

::::::::::::::::

```bash
mkdir rnaseq
(
  cd rnaseq
  wget https://zenodo.org/record/4541751/files/GSM461177_1_subsampled.fastqsanger
  wget https://zenodo.org/record/4541751/files/GSM461177_2_subsampled.fastqsanger
  wget https://zenodo.org/record/4541751/files/GSM461180_1_subsampled.fastqsanger
  wget https://zenodo.org/record/4541751/files/GSM461180_2_subsampled.fastqsanger
  wget https://zenodo.org/record/4541751/files/Drosophila_melanogaster.BDGP6.87.gtf
  wget https://hgdownload.soe.ucsc.edu/goldenPath/dm6/bigZips/dm6.fa.gz
  gunzip dm6.fa.gz  # STAR index requires an uncompressed reference genome
)
```

### STAR Genome index

To run the STAR aligner tool, index files generated from the reference genome are needed.
The index files can be downloaded, or generated yourself from the unindexed reference genome.

These two options are detailed below -- choose the one most appropriate to your set-up.

:::::::::::::::::::: spoiler

#### Option 1: Download the index

At least 9 GB of memory is required to generate the index, which will occupy 3.3GB of disk.

If your computer doesn't have that much memory, then you can download the directory
by running the following in the `rnaseq` directory:

```bash
wget https://dataverse.nl/api/access/datafile/266295 --output-document - | tar -C rnaseq -x --xz
```

:::::::::::::::::::::

::::::::::::::::::::: spoiler

#### Option 2: Generate the index yourself

To generate the genome index yourself: create a new file named `dm6-star-index.yaml`
in the the `novice-tutorial-exercises` directory with the following contents:

```yaml
InputFiles:
 - class: File
   location: rnaseq/dm6.fa
   format: http://edamontology.org/format_1929  # FASTA
IndexName: 'dm6-STAR-index'
Overhang: 36
Gtf:
 class: File
 location: rnaseq/Drosophila_melanogaster.BDGP6.87.gtf
```

Next use the CWL reference runner `cwltool` that you installed above and the CWL description
for the indexing mode of the STAR aligner that was downloaded in the `bio-cwl-tools` directory
to index the genome and place the result in the `rnaseq` directory alongside the other files:

```
cwltool --outdir rnaseq/ bio-cwl-tools/STAR/STAR-Index.cwl dm6-star-index.yaml
```

It should take 10-15 minutes for the index to be generated.

::::::::::::::::::::: callout

### STAR Index Memory Requirements
To generate the genome index you will need at least 9 GB of RAM.

If you do not allocate enough RAM the tool will not crash, but the process will stick
on the following step:
```
... sorting Suffix Array chunks and saving them to disk...
```
If this step does not finish within 10 minutes then it is likely the process has failed,
and should be cancelled.

MacOS users can configure  Docker Desktop to allocate more memory
from the menu "Preferences" and then selecting "Resources".

Windows users can configure WSL 2 to allocate more memory by opening the PowerShell
and entering the following:

```bash
# turn off all wsl instances such as docker-desktop
wsl --shutdown

notepad "$env:USERPROFILE/.wslconfig"
```

In `.wslconfig` add the following

```text
[wsl2]
memory=9GB
```

Save the file and right-click the Docker icon in the notifications area (or System tray)
and then click "Restart Dockerâ€¦"

:::::::::::::::::::::

:::::::::::::::::::::
